MSC4143: MatrixRTC – Real-time communication over Matrix (#4143)
matrix-org/matrix-doc: proposals/4143-matrix-rtc.md

Reviewer: Implementation requirements:

- [x] Client
- [x] Server

Reviewer: This proposal and its implementations have an extensive history that went through multiple iterations. Current implementations are linked below using pull requests wherever possible. In some cases though, direct links into the source code were more practical and have been used instead.

To test MatrixRTC with the currently implemented versions of https://github.com/matrix-org/matrix-spec-proposals/pull/4196 and https://github.com/matrix-org/matrix-spec-proposals/pull/4195, you need a homeserver that has implemented and enabled https://github.com/matrix-org/matrix-spec-proposals/pull/4140 and https://github.com/matrix-org/matrix-spec-proposals/pull/4354. The server also needs to have https://github.com/element-hq/lk-jwt-service and a LiveKit SFU configured and advertised. Within the developer options of Element Call you can then force the Matrix 2.0 mode.

### Slots

- JS SDK: https://github.com/matrix-org/matrix-js-sdk/pull/5445
- Element Web: https://github.com/element-hq/element-web/pull/34392

### Membership

- JS SDK
  - Event schema & validation
    - [`rtc.ts`](https://github.com/matrix-org/matrix-js-sdk/blob/fb95ae8f34a3aad30f129feb189aa0cf18ef5733/src/matrixrtc/membershipData/rtc.ts)
    - https://github.com/matrix-org/matrix-js-sdk/pull/5423
  - Join/update/leave lifecycle
  	- [`MembershipManager.ts`](https://github.com/matrix-org/matrix-js-sdk/blob/fb95ae8f34a3aad30f129feb189aa0cf18ef5733/src/matrixrtc/MembershipManager.ts)
  	- https://github.com/matrix-org/matrix-js-sdk/pull/5437
  - Joined-member conditions & session computation: [`isValidMembership`](github.com/matrix-org/matrix-js-sdk/blob/fb95ae8f34a3aad30f129feb189aa0cf18ef5733/src/matrixrtc/MatrixRTCSession.ts#L875)
  - Ephemeral sticky events map: [`room-sticky-events.ts`](https://github.com/matrix-org/matrix-js-sdk/blob/fb95ae8f34a3aad30f129feb189aa0cf18ef5733/src/models/room-sticky-events.ts)
  - Element Call join path: [`enterRTCSession`](https://github.com/element-hq/element-call/blob/e20abb19fbdb2b6fd729e02ab66b53c70fec48d0/src/state/CallViewModel/localMember/LocalMember.ts#L791) (sticky membership is enabled in `MatrixRTCMode.Matrix_2_0`)

> [!NOTE]
> The JS SDK still uses the properties `user_id` and `device_id` in the `member` object of `m.rtc.member` events. These have been eliminated in the MSC as they can be deduced by other / more secure means.
>
> Similarly, the SDK still includes the `versions` property and `m.reference` relations when sending `m.rtc.member` events. These have been removed from the MSC as they were unneeded.
>
> The `member.membership` property of `m.rtc.member` events is not implement. Instead the JS SDK treats the absence of the `application` object as implicitly left.
>
> Equally, the `leave_reason` code `slot_closed` has not yet been implemented. It blocks on the combination of https://github.com/matrix-org/matrix-js-sdk/pull/5445 and https://github.com/matrix-org/matrix-js-sdk/pull/5437.

### Discovery of transport infrastructure
 
- Server: https://github.com/element-hq/synapse/pull/18967
- JS SDK: https://github.com/matrix-org/matrix-js-sdk/pull/5104
- Element Call: [`RtcTransportAutoDiscovery.ts`](https://github.com/element-hq/element-call/blob/e20abb19fbdb2b6fd729e02ab66b53c70fec48d0/src/state/CallViewModel/localMember/RtcTransportAutoDiscovery.ts)
- Element Web: [`CallStore.fetchTransports`](https://github.com/element-hq/element-web/blob/5bcfe5aee656937b068b867230802e2eda2f7c0b/apps/web/src/stores/CallStore.ts#L60)

> [!NOTE]
> When ran in widget mode, Element Call cannot yet use the `/transports` endpoint because the widget API doesn't expose access to it. This is covered by https://github.com/matrix-org/matrix-spec-proposals/pull/4515.

### End-to-end encryption

- JS SDK
  - Key generation & rotation: [`RTCEncryptionManager.ts`](https://github.com/matrix-org/matrix-js-sdk/blob/fb95ae8f34a3aad30f129feb189aa0cf18ef5733/src/matrixrtc/RTCEncryptionManager.ts)
  - Key distribution via to-device messages: [`ToDeviceKeyTransport.ts`](https://github.com/matrix-org/matrix-js-sdk/blob/fb95ae8f34a3aad30f129feb189aa0cf18ef5733/src/matrixrtc/ToDeviceKeyTransport.ts)
- Element Call
  - Bridging key into LiveKit transport: [`matrixKeyProvider.ts`](https://github.com/element-hq/element-call/blob/e20abb19fbdb2b6fd729e02ab66b53c70fec48d0/src/e2ee/matrixKeyProvider.ts)

> [!NOTE]
> The JS SDK still relies on the (claimed) `device_id` property in `m.rtc.member` events while the MSC has eliminated `claimed_device_id` and instead infers the device ID from authenticated properties obtained in the decryption process of `m.rtc.member` events.
>
> Instead of `m.rtc.encryption_key`, the JS SDK uses the legacy event type `io.element.call.encryption_keys` with a slightly different but equivalent schema.
>
> Additionally, the JS SDK generates 16-byte instead 32-byte keys.
>
> Discarding unencrypted to-device key messages and verifying the to-device message sender and device ID against the `m.rtc.member` event are not yet implemented in the JS SDK. These block on changes to the widget API (https://github.com/element-hq/voip-internal/issues/651 / https://github.com/element-hq/voip-internal/issues/655).

RTC applications by default.

### MatrixRTC Membership

A MatrixRTC membership is represented by a sticky `m.rtc.member` event
([MSC4354](https://github.com/matrix-org/matrix-spec-proposals/pull/4354) Sticky Events). These
events describe a participant’s presence in an MatrixRTC slot and provide sufficient metadata for
other room members to detect and join the same slot. 

The membership events are implemented as a key-value store according to the addendum section of
[MSC4354 Sticky Events](https://github.com/matrix-org/matrix-spec-proposals/pull/4354) giving the
same semantics as Matrix room state. If the room is encrypted `m.rtc.member` events are also
encrypted and vice versa.

An `m.rtc.member` sticky event can be either **Connected** or **Disconnected**

* **Connected**, if  
  * **Slot open**: the `sticky_key` of the event needs to match the `member.id` AND the `slot_id`
    matches the state\_key of the slot event.  
  * **Content:** MUST match the JSON content schema for connecting to a slot (see below).  
  * The sender is still a member of the room (not kicked / left)  
  * **Event is sticky**: The sticky event needs not to be expired as described in [MSC4354 Sticky

>     Events](https://github.com/matrix-org/matrix-spec-proposals/pull/4354)  

Reviewer: We should clarify that this is to ensure a consistent membership view across all participants. Continuing to consider calls connected after stickyness has exceeded would mean that some participants might not have received the membership event and thus would not see the participant.

Reviewer: Clarified in 9852003412d1557d474503bc6f4b50bab5e47c4a.

  allowing clients to avoid connecting to participants outside their area of interest.  
  **Note** application-specific metadata as part of the MatrixRTC slot `application` object have
  precedence over the MatrixRTC member ones.
* `m.relates_to` —  The `m.relates_to` field optionally references the initial connect event,
  distinguishing connect events from updates. Clients SHOULD include this field when sending updates
  to an existing `m.rtc.member` event to ensure continuity in membership lifecycle tracking, enable
  accurate historical reconstruction, and allow deriving the participant’s start time using the
  `origin_server_ts` of the referenced event.  
* `member` — Uniquely identifies this participation instance; includes:  
  * `id` — UUID to distinguish multiple participations, even for the same user and same device. This
    ID can also serve as a canonical identifier for certain MatrixRTC transports, helping to prevent
    metadata leakage if used as additional entropy, e.g., `SHA-256(user_id|device_id|member.id)`. 
    The ID MUST be unique for each connect event.
  * `claimed_device_id` — Matrix device identifier.  

>   * `claimed_user_id` — Matrix user ID.  

Reviewer: Stupid question: Why do these use the `claimed_` prefix?

Reviewer: Clarified out of band, this is to emphasize that these values are indeed "claimed" and not guaranteed to be correct. Recipients should verify them.

Reviewer: Clarified in 57c8cbe2a3a0b14163b46aa0dc1894e1c7f14267.

> # MSC4143: MatrixRTC

Reviewer: Given that this is an inline comment and, therefore, buried in the history of this proposals changes, I'll resolve this. I have added a replacement file comment here: https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3644032580

sufficient `power_level`. Clients MUST react on and respect latest `m.rtc.slot` definitions as
defined by the Matrix room state at all times. The ability to open and close slots with different
patterns enables a wide variety of use cases, from always-on shared spaces to short lived scheduled
meetings.

Slots are named by the RTC app, but are expected to typically be ordinals (e.g. equivalent to line 1
or line 2 on a telephone). If the app needs to define its slots out of band (e.g. mapping them to
widget IDs) then it can use those IDs as names. However, given a slot is the mechanism around which
sessions converge, it must have a predictable name. Unpredictable IDs such as session IDs should
never be used to name a slot.

#### Opening a MatrixRTC Slot

A slot is opened by sending an `m.rtc.slot` state event with `state_key = slot_id`. This event
authorises MatrixRTC members that intend to participate in the slot and follows the schema below:

```json5
// Example: an open slot with application-specific metadata
{
  "application": {
    "type": "m.call",
    // optional: app specific slot metadata

>     "m.call.id": UUID,           // Note your application must handle rollback due to state resolution

Reviewer: I _think_ this may just refer to letting whatever the latest resolved value is take precedence?

There's this example further down:

> * `id_state_res` This `m.rtc.slot` state event is subject to Matrix state resolution. Hence, it is
  divided potentially into two intervals. If a rollback occurs during an active MatrixRTC session,
  application logic MUST handle conflicts (e.g., a rolled-back UUID).

So if you start a call with some UUID A but then state resolution rolls that state event back to another one with UUID B, then your client is supposed to switch you to the call with UUID B? 🤔

>     "m.call.id": UUID,           // Note your application must handle rollback due to state resolution

Reviewer: Clarified out of band, this interpretation appears to be correct.

#### RTC Transports in `m.rtc.member` Events

Each `m.rtc.member` event contains an `rtc_transports` array, which allows other clients to
determine how to connect to, and exchange real-time data with that participant. This array includes
at least one transport JSON object, each containing at least the following required field:

* `type` (string) — Identifies the RTC transport.  
* Additional fields may be present depending on the RTC transport `type`.

The `versions` field of the `m.rtc.member` event SHOULD be used to determine which transports are
supported by all participants. Clients SHOULD publish their RTC data using at least one transport
supported by their application, constrained to the set of transports supported by all participants.
Clients MAY additionally publish their media via multiple transports.

To render a MatrixRTC session, a client MUST be prepared to connect to as many transports as there
are members currently connected. For bandwidth efficiency, clients are recommended to subscribe to
only one transport per member.

The exact procedures for subscribing to and publishing real-time data are defined in the dedicated
MSCs for each transport type.

> #### Discovery of RTC Transports

Reviewer 2: This section doesn't say whether the endpoint should be rate-limited, and whether it requires authentication.

Currently the implementation in Synapse requires authentication.

Reviewer 3: I think it's expected to require auth so we should update this section. Also, I'm assuming we *would* want guests to be able to access it.

I don't think this needs special ratelimiting guidance, it's only returning configuration.

Reviewer 4: It would be good for the MSC to explicitly say the stance on authentication. A PR came in to Synapse removing the authentication on this endpoint: https://github.com/element-hq/synapse/pull/19646#discussion_r3065623704

As it stands I am personally leaning to 'requires authentication' but I can see why people might believe this should be public, so especially worthwhile to define it clearly.

Reviewer 5: I agree, it should stay authenticated. We will clarify that in the MSC

Reviewer: It looks like this was resolved in https://github.com/matrix-org/matrix-spec-proposals/commit/866ca090e95e17b82da4614c23e9cdd24aa5466e.

Future constraint definitions would likely exist alongside the `"application"` key within the
`m.rtc.slot` event structure.

## Security considerations

### Discoverability of RTC Infrastructure

RTC infrastructure details are disseminated to all participants through `m.rtc.member` events. This
transparency means that anyone part of a MatrixRTC session can view and understand the
infrastructure, which could potentially lead to unauthorized resource use. However, each
infrastructure type defines its own authentication mechanisms, as detailed in its specific MSC.
These mechanisms may involve a service interacting with the homeserver to determine whether a user
is authorized to utilize the infrastructure.

### Forward secrecy for end-to-end encryption of media streams

The considerations to ensure forward secrecy are described in the [Key
Distribution](#key-distribution) section above.

### End-to-end media encryption key rotation lag

The proposed key rotation semantics does mean that a participant could continue to decrypt media
that was sent in the three seconds after leaving the session.

> ## Unstable prefix

Reviewer 4: Should this MSC have an advertisement in `/versions`?

A PR came in to Synapse adding one: https://github.com/element-hq/synapse/pull/19646#discussion_r3065617589

but currently the MSC doesn't prescribe one (which it should if it needs one, + consider the `org.matrix.msc4143.stable` variant for later down the line)

Reviewer: Yes, definitely makes sense. Have added both version entries in 6bc4ff09f95f05c2fc4e958534b37b80c1e35004.

* How different RTC streams are interpreted.  
* Which MatrixRTC transport types are supported or required.

For example, a [Third Room](https://thirdroom.io) like experience could include information about
the virtual scene, e.g., a place, available objects or weather conditions.

This modular design makes MatrixRTC flexible. For example, a Jitsi-based conference could be
supported by defining a new application type and a corresponding RTC transport. While such a session
would not be compatible with Matrix clients that do not implement the Jitsi transport, those clients
would still be able to detect the presence of an unsupported application and transport type, and
handle it with a sensible user interface.

The minimum Application definition consists of a simple JSON object

```json5
{
  "application": {
    "type": "m.call", // The # character MUST NOT appear in a valid type string.
    // Optional: application-specific metadata
    "m.call.spatial_audio": true
  }
}
```

The `type` field MUST be a string that does not contain the `#` character. The type field SHOULD
follow the [*Common Namespaced Identifier

> Grammar*](https://spec.matrix.org/v1.16/appendices/#common-namespaced-identifier-grammar).

Reviewer: Would there be any downsides to enforcing the common namespaced identifier grammer with MUST here? That excludes `#` automatically and avoids name clashing.

Reviewer: Clarified out of band, we should probably just make this a MUST.

Reviewer: Clarified in 118c9dbe0c33427410badd5cf5c1e4c7837534ec.

Each application type MUST have its own MSC, which specifies the additional fields and defines how

> communication with the corresponding transport is handled. For example:

Reviewer: I think MUST doesn't work well here. The specification consists of requirements that are applied to implementations but implementations don't write MSCs. I believe what we mean to say here is that concrete applications types are introduced in separate proposals rather than contained in this one?

```suggestion
Each application type specifies the additional fields and defines how communication with the
corresponding transport is handled. Concrete applications types are introduced in separate proposals
such as:
```

* `m.call` — voice and video calling, as defined in
  [MSC4196](https://github.com/matrix-org/matrix-spec-proposals/pull/4196).

To facilitate interoperability, ideally each application type should provide a Matrix widget that
can serve as a reference implementation for clients. This allows client developers to support new
application types by embedding or integrating the widget, without having to implement the full

> application logic themselves.

Reviewer: This paragraph feels a bit tricky. Application types are just a set of requirements and semantics that we'll write into the spec. So an application type itself cannot "provide a Matrix widget". I think this is something that we would like creators of new application types to do?

That would make it a [spec change process](https://spec.matrix.org/proposals/) thing. It could be an additional implementation requirement for application type proposals, for instance. Even that feels difficult though because widgets are themselves not currently part of the spec. So a requirement like this would block this proposal on standardizing widgets.

I think ideally we should decouple any references to widgets from this proposal and just omit the entire paragraph.

```suggestion
```

### MatrixRTC Slot and Constraining Slots

A **MatrixRTC slot** is the container for MatrixRTC members and temporal overlapping MatrixRTC
members form a MatrixRTC session. Each slot is identified by a unique `slot_id`, tied to a specific
`application` type, and represented by the `m.rtc.slot` state event with the `slot_id` as the state
key. 

> [!NOTE] Conceptually, a slot is like a virtual meeting room: MatrixRTC sessions occur within

> > slots, and multiple sessions may occur consecutively without changing the slot’s configuration.

Reviewer: Looks like the `[!NOTE]` needs to go on its own line for the [alert](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts) to render properly.

```suggestion
> [!NOTE]
> Conceptually, a slot is like a virtual meeting room: MatrixRTC sessions occur within
> slots, and multiple sessions may occur consecutively without changing the slot’s configuration.
```

Slots are part of the shared room state and can only be created or modified by authorized users with

> sufficient `power_level`. Clients MUST react on and respect latest `m.rtc.slot` definitions as

Reviewer: `power_level` is slightly ambiguous here because this isn't actually an identifier (the state event is ` m.room.power_levels`). I think just using the plain words "power level" is a bit clearer.

```suggestion
Slots are part of the shared room state and can only be created or modified by authorized users with
sufficient power level. Clients MUST react on and respect latest `m.rtc.slot` definitions as
```

defined by the Matrix room state at all times. The ability to open and close slots with different
patterns enables a wide variety of use cases, from always-on shared spaces to short lived scheduled
meetings.

Slots are named by the RTC app, but are expected to typically be ordinals (e.g. equivalent to line 1
or line 2 on a telephone). If the app needs to define its slots out of band (e.g. mapping them to
widget IDs) then it can use those IDs as names. However, given a slot is the mechanism around which
sessions converge, it must have a predictable name. Unpredictable IDs such as session IDs should

> never be used to name a slot.

Reviewer: The term "slot name" confused me initially because I first thought this is something separate from the `slot_id`. It's only clarified further down that both are in fact the `state_key` of the `m.rtc.slot` event.

> * **state key**: The state key of the `m.rtc.slot` state event, referred to as the `slot_id`, serves as the slot name. 

Still later on, it is revealed that `slot_id`s are themselves composed:

> {application.type}#{application_slot_id}

I think this paragraph mainly describes the `application_slot_id` suffix?

In either case, since the terms and grammar are defined in the section below, maybe this paragraph should be moved from here and to [the section below](https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3302470915) to avoid ambiguity?

```suggestion
```

#### Opening a MatrixRTC Slot

A slot is opened by sending an `m.rtc.slot` state event with `state_key = slot_id`. This event
authorises MatrixRTC members that intend to participate in the slot and follows the schema below:

```json5
// Example: an open slot with application-specific metadata
{
  "application": {
    "type": "m.call",
    // optional: app specific slot metadata
    "m.call.id": UUID,           // Note your application must handle rollback due to state resolution
    "m.call.voice_only": true
  }
}

state_key: "m.call#ROOM"         // slot_id
```

**Field description:**

* **state key**: The state key of the `m.rtc.slot` state event, referred to as the `slot_id`, serves

>   as the slot name.  

Reviewer: Picking up on [my comment above](https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3302438560), it looks like we're not assigning further meaning to the term "slot name". Given that we already have `state_key` and `slot_id` as terms, maybe we could omit "slot name" entirely to avoid potential ambiguity?

```suggestion
* **state key**: The state key of the `m.rtc.slot` state event, referred to as the `slot_id`.
```

* `application` An application JSON object, which **MUST** specify the application type and MAY
  include additional fields which **constrain** the application (e.g., restricting a call to be
  voice-only). **Note** those additional fields can be used in combination with MatrixRTC member

>   ones subject to the application specific MSC.

Reviewer: I was a little confused at first because the member events haven't yet been introduced here and I didn't understand that they have an `application` block, too. Maybe we could spell this out a little more?

```suggestion
  voice-only). **Note** those additional fields may be extended through fields in the `application`
  content block of `m.rtc.member` events. If a field occurs in both, the value from the `m.rtc.slot` event
  takes precedence.
```

The `slot_id` of an open slot acts like a virtual address where participants are allowed to meet.
The grammar for the `slot_id` is a well formed state key confined such that members of different
MatrixRTC applications never occupy the same slot according to: 

```json5
{application.type}#{application_slot_id}
```

Where

* `application.type` is the `type` field in the application JSON object  
* The `#` character MUST NOT be used in either `application.type` or `application_slot_id`  
* `application_slot_id` is the application-specific slot ID. Each application MSC defines its own
  schema (e.g., `ROOM`, `1`, `2`) to allow multiple parallel slots of the same type according to the

>   application requirements.

Reviewer: Picking up on [my comment above](https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3302438560) and reintroducing the bits removed there here:

```suggestion
* `application_slot_id` is the application-specific slot ID. Each application defines its own
  schema (e.g., `ROOM`, `1`, `2`) to allow multiple parallel slots of the same type according to the
  application requirements. If the application needs to define its slots out of band (e.g. mapping them to
  widget IDs) then it can use those IDs as `application_slot_id`. However, given a slot is the mechanism
  around which sessions converge, it MUST have a predictable `application_slot_id`.
```

m.rtc.slot[id_2] {content_2}                         ███████████████████

m.rtc.slot[id_state_res] {content_ABC}         ████XXXXXXXXXXXXXXXXXXXXXXXXXXXXX
m.rtc.slot[id_state_res] {content_XYZ}             █████████████████████████████
                                                   |
                                                   |
                                                   Rollback due to Matrix state resolution

Time                                   ────────────────────────────────────────────────────►
```

* `id_1` comprises a long-lived slot suited for a Discord-style experience or a shared whiteboard
  where people can hop on and hop off as desired.  
* `id_2` suited for a scheduled conference meeting, where the state event might be managed using
  cancellable delayed events (MSC4140) or a bot.  
* `id_state_res` This `m.rtc.slot` state event is subject to Matrix state resolution. Hence, it is
  divided potentially into two intervals. If a rollback occurs during an active MatrixRTC session,
  application logic MUST handle conflicts (e.g., a rolled-back UUID).

#### Recommended User Experience

> The lifecycle of a slot follows four distinct states:

Reviewer: I think it's only three states (closed, open+active and open+inactive)?

```suggestion
The lifecycle of a slot follows three distinct states:
```

```
Closed
Open
 ├─ Active   --- at least one m.rtc.member is connected
 └─ Inactive --- no member is connected
```

Slots may transition between these states as a slot is opened or closed, or as users connect or disconnect.

While **participation** in a MatrixRTC session is possible if a **slot is open**, **management** of
slots depends on the user’s **power level**. Clients MUST respect these distinctions in the user
experience.

For **open** slots in a room, clients SHOULD clearly represent the current slot’s activity state:

* **Active:** clients SHOULD indicate which slots are currently active in a room (e.g., by showing
  participant count, showing an icon, list entry, or join button).   
* **Inactive:** clients SHOULD indicate that the slot is available and provide a way for a user to
  join as the first member (e.g., via a “Call” button or placeholder icon).

**Slot management** UX SHOULD only exposed to users who have sufficient power level to manage slots:

* **Closed slot:** The UX SHOULD provide controls to manage the slot (e.g., opening or modifying the
  slot)  
* **Open slot:** The UX SHOULD provide controls to perform management (e.g., closing or modifying

>   the slot)

Reviewer: I think apart from a small number of exceptions, the spec normally doesn't define any UX requirements to give maximum freedom to clients. Power levels are enforced on the server. The rest feels like sensible things a client could do but I'm not sure if all of this should be spelled out using normative language?

As a possibly extreme option, I think we could just omit all the UX references:

```suggestion
While **participation** in a MatrixRTC session is possible if a **slot is open**, **management** of
slots depends on the user’s **power level**. 
```

Or as a middle ground, we could downgrade the normative language to just represent examples:

```suggestion
While **participation** in a MatrixRTC session is possible if a **slot is open**, **management** of
slots depends on the user’s **power level**. Clients should be aware of this distinction when constructing
their user interfaces.

For **open** and **active** slots, for instance, clients could display the participant count, an icon,
a list entry, or join button. For **open** but **inactive** slots, in turn, clients could indicate that the
slot is available and provide a way for a user to join as the first member (e.g., via a “Call” button or
placeholder icon). Finally, for users who have sufficient power level to manage slots, clients could
display a slot management UI with controls to open, close or modify slots.
```

Multiple active slots may exist in the same room if the application type supports them.

Clients MAY present an option to room administrators to enable specific applications in the room by
adding a slot and a type to room state.  Future MSCs may change the defaults for new rooms to enable
RTC applications by default.

### MatrixRTC Membership

A MatrixRTC membership is represented by a sticky `m.rtc.member` event
([MSC4354](https://github.com/matrix-org/matrix-spec-proposals/pull/4354) Sticky Events). These
events describe a participant’s presence in an MatrixRTC slot and provide sufficient metadata for
other room members to detect and join the same slot. 

The membership events are implemented as a key-value store according to the addendum section of
[MSC4354 Sticky Events](https://github.com/matrix-org/matrix-spec-proposals/pull/4354) giving the
same semantics as Matrix room state. If the room is encrypted `m.rtc.member` events are also

> encrypted and vice versa.

Reviewer: Mirroring [my comment](https://github.com/matrix-org/matrix-spec-proposals/pull/4354/changes#r3301585363) on #4354, we may want to consider moving the map algorithm here as it seems rather feature-specific to MatrixRTC.

Reviewer: The map algorithm with stay in #4354 as discussed in the thread linked above.

  "versions": [
    "v0",
    "example.mscXXXX.asymmetric_encryption"
  ],
}
```

**Field explanations:**

* `slot_id` — The slot this member belongs to.  
* `application` — Is a JSON object that identifies the MatrixRTC application type; Besides the
  mandatory `type` field which needs to match the MatrixRTC slot identified by `slot_id` it **may**
  include application-specific metadata (see MatrixRTC Application Types). For example, a [Third
  Room](https://thirdroom.io) style experience could include approximate position data on a map,
  allowing clients to avoid connecting to participants outside their area of interest.  
  **Note** application-specific metadata as part of the MatrixRTC slot `application` object have
  precedence over the MatrixRTC member ones.
* `m.relates_to` —  The `m.relates_to` field optionally references the initial connect event,
  distinguishing connect events from updates. Clients SHOULD include this field when sending updates
  to an existing `m.rtc.member` event to ensure continuity in membership lifecycle tracking, enable
  accurate historical reconstruction, and allow deriving the participant’s start time using the

>   `origin_server_ts` of the referenced event.  

Reviewer: Updates to membership events seem to have replacement semantics. I'm curious why they use `m.reference` rather than `m.replace`?

Reviewer: Resolving in favor of https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3303321664.

* `member` — Uniquely identifies this participation instance; includes:  
  * `id` — UUID to distinguish multiple participations, even for the same user and same device. This
    ID can also serve as a canonical identifier for certain MatrixRTC transports, helping to prevent
    metadata leakage if used as additional entropy, e.g., `SHA-256(user_id|device_id|member.id)`. 

>     The ID MUST be unique for each connect event.

Reviewer: Maybe we should make the hashing normative with MUST or at least SHOULD to reliably prevent that metadata leakage?

Also related, should these be identical for the same device when joining, leaving and rejoining the same slot?

Reviewer: Actually the `|`-concatenation is vulnerable to delimiter injection which can produce hash collisions. We should probably recommend the canonical JSON array alternative that is also used in https://github.com/matrix-org/matrix-spec-proposals/pull/4195.

Reviewer: Fixed in 3ccd90b57ab4a36b2cf80a27f577a46f5540be25.

- `m.relates_to`: Clients SHOULD include the `m.relates_to` field referencing the original
  `m.rtc.member` join event when disconnecting. This facilitates accurate session history and
  retrospective computations by fetching the relation to determine the participant’s original join
  time or associated metadata.  
- `sticky_key:` The sticky key from the disconnecting MatrixRTC member.  
- `disconnect_reason`: The `disconnect_reason` object is **optional** and provides additional
  context when a participant disconnects from a call. It is only meaningful if the user has
  **previously attempted to connect** (i.e., has sent at least one valid `m.rtc.member` event for
  the slot) This ensures that the disconnection reason refers to a real connection lifecycle rather
  than a pre-join cancellation. 

`disconnect_reason` **Field explanations:**

| Field | Type | Required | Description |
| ----- | ----- | ----- | ----- |
| `class` | string | ✅ | High-level category of the disconnection or error. |
| `reason` | string | ✅ | Machine-readable identifier of the specific cause. |
| `description` | string | ⚪ | Optional human-readable explanation providing additional context. |

> **Class categories and examples:**

Reviewer: The table below appears to also exist in #4196 and maybe it belongs there rather than here?

Additionally it's unclear if the reasons really meant to just be examples or if we want them to be normative.

Reviewer: Clarified out of band, this needs further thought as it is unclear why we need such a sophisticated connection error system.

Reviewer: It looks like this originated from a desire to be able to represent SIP's error cases: https://en.wikipedia.org/wiki/List_of_SIP_response_codes

Reviewer: All addressed in d597297b3b380b7eaa9bf17766600513a99ed05a.

> | Class | Example Reason | Description / When Used |

Reviewer: Are the reasons really meant to just be examples or do we want them to be normative?

Reviewer: Resolving in favour of https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3311652712.

* **Precise membership tracking** — RTC applications rely on a high-resolution understanding of the
  lifetime of each `m.rtc.member` object. Knowing exactly when a participant joins or leaves is
  critical for computing session state.  
* **Handling (unintentionally) disconnected clients** — A client that loses network connectivity
  must not be treated as continuously connected. Otherwise, the session state may become
  inconsistent or appear to include phantom participants.

Clients MUST handle network disconnects or crashes by sending empty leave events once reconnection
is possible, or by using cancellable delayed leave events
([MSC4140](https://github.com/matrix-org/matrix-spec-proposals/pull/4140)).

Clients SHOULD use cancellable delayed events to implement a “deadman switch” for MatrixRTC
membership by sending a leave event ahead of the join event as a delayed event with a reasonable
delay (e.g., 15–30 seconds). The client then periodically resets the delayed event’s timer. If the
timer expires due to a missing reset, the leave event is automatically emitted, marking the
participant as disconnected and ensuring accurate session state even in cases of sudden

> disconnection, crashes, or network failures.

Reviewer: I wonder if we should make the dead-man's-switch mandatory given that we can expect the alternative of sending the leave on re-connect to be poor UX?

Reviewer: On second thought SHOULD seems sufficient.

m.rtc.member[0]            |████████████████████████   ████████████████████████|████
                           |^ connect   disconnect ^   ^ connect   disconnect ^|
                           |                                                   |
m.rtc.member[1]            |       ███████████████████████████████████████     |
                           |       ^ connect                  disconnect ^     |
                           |                                                   |
m.rtc.member[2]            |            ████████████████████████               |
                           |            ^ connect   disconnect ^               |
                           |                                                   |
Slot (open)          [*********************************************************]
                           |                                                   |
Session lifetime           [***************************************************]

Time               ───────────────────────────────────────────────────────────────────►
```

The lifetime of a MatrixRTC session is implicitly defined by the union of all **connected** member

> intervals, bounded by the slot’s open and close times:

Reviewer: How about a case like this when there is an interval in between where nobody is connected. Are these two sessions or just one?
```
                           Session Start                                       Session End
                               |                                                   |
    m.rtc.member[0]            |████████████████████████   ████████████████████████|
                               |^ connect   disconnect ^   ^ connect   disconnect ^|
                               |                                                   |
    m.rtc.member[1]            |████████████████████████   ████████████████████████|
                               |^ connect   disconnect ^   ^ connect   disconnect ^|
                               |                                                   |
    Slot (open)          [**************************************************************]
                               |                                                   |
    Session lifetime           [***************************************************]

    
    Time                   ───────────────────────────────────────────────────────────────────►

```

If I understood the text below correctly those should be two sessions?

Reviewer: Clarified out of band: Yes.

session and converge on the session that wins the race.

If conflicts occur later during a session, the better user experience may be to treat them as two
concurrent calls within the same room, allowing users to choose which session to join. In practice,
this could be implemented as a prompt (e.g., *“There’s already a call happening — do you want to
join it?”*), potentially automated to accept by default when no other participants remain in the
local session.

#### Session history

As MatrixRTC does not emit a single event describing past sessions, historic MatrixRTC sessions MUST
be reconstructed from room history. Clients SHOULD cache reconstructed sessions to avoid repeated,
expensive history pagination.

In general, all required information is present as part of the Matrix DAG:

* **Sticky events** are persistently recorded in the Matrix DAG.  
* **`m.rtc.slot` state event transitions**, even when affected by state resolution, are visible as
  part of the timeline.

Counterintuitively, **state resolution is not a problem** — RTC sessions occurred in real life, and
their existence remains valid regardless of how the room state later resolves.

> 

Reviewer: This paragraph appears to be duplicated in the "Potential Issues" section below.
```suggestion
```

However, for this reconstruction process to work **reliably and deterministically**, the Matrix 
protocol requires **consistent message ordering (on a homeserver basis)** across both state and 
timeline events. This remains an open problem and is discussed further in the 
[*Impact of Message Ordering wrt. Session History*](#impact-of-message-ordering-wrt-session-history) 
section. As a temporary workaround, the reconstruction process uses `origin_server_ts` ordering.

General concept
* Paginate backward through room history to collect relevant events:  
  * `m.rtc.slot` events (slot open/close boundaries)  
  * `m.rtc.member` events (connects, updates, disconnects)  
* Correlate `m.rtc.slot` events with surrounding `m.rtc.member` events to reconstruct membership

>   intervals.  

Reviewer: I wonder if this is going to be cumbersome especially since slots and member events have no `m.relates_to` connection that could speed up the look-up?

When the poll history archive was first added in Element's clients, the initial solution used back-pagination to find all the poll start events and then `/relations` to find votes belonging to each event. I remember this being painfully slow and it even had the benefit of using relations for some part. Performance was later improved by anchoring all poll start events on an empty state event via https://github.com/matrix-org/matrix-spec-proposals/pull/4013.

Maybe we should consider adding a relation from each member's first event in a session back to the slot event? Once you have the slot event, you could then find all connect events using `/relations`. If we then also use `m.replace` rather than `m.reference` to link the chain of membership events, then the server should include the latest replacement event for each connect event (which is the disconnect event) as a bundled aggregation. In other words, given a slot event, a single call to `/relations` would give us all we need to build session history without the need to paginate all history.

Reviewer: This needs further thought with respect to what we actually want to use the relations to.

Relating to the slot event is tricky because we would have to update the relation if the slot changes due to state resolution.

Furthermore, the bundled aggregation of both `m.replace` and `m.reference` isn't very helpful in our case. By the time you scroll up to the connect event, you will usually have come past all the related events already. Even if you didn't, it seems unclear what clients would do with the relations chain if / when rendering the connect event.

One plausible use case is a per-room call archive as in a tab inside a room that lists past calls. This would benefit from something akin to [`GET /_matrix/client/v1/rooms/{roomId}/threads`](https://spec.matrix.org/v1.18/client-server-api/#querying-threads-in-a-room:~:text=API%3A-,GET%20%2F%5Fmatrix%2Fclient%2Fv1%2Frooms%2F%7BroomId%7D%2Fthreads) for fetching all RTC member events in a room. This in turn would require a dedicated relation type, however.

Reviewer: A major drawback of a dedicated relation type is that the type is leaked to the server. The benefit is one of performance only. Therefore, we're currently leaning towards not switching to a dedicated relation type.

Switching from `m.replace` to `m.reference` is still being discussed. It doesn't only change `rel_type` though since the updated content needs to go into `m.new_content`. The top-level fields in `content` are supposed to act as fallback for clients that don't support replacements. The only thinkable fallback in our case would be to just repeat the updated properties. That'll mean we have them in both `content` and `m.new_content`, however. Alternatively we could claim that there is no fallback and just omit anything that is not `m.new_content` or `m.relates_to` from `content`.

Reviewer: We've concluded to not use relations at all initially. We may introduce them later if we find a good use case for them.

Reviewer: Fixed in a5bef33503fb91ffd038e6258a9cac4b364c1a9e.

Transports describe how clients connect to peers or servers, manage connections, publish and
subscribe to media streams. A transport can represent different RTC backends such as SFUs, MCUs, or
full-mesh peer-to-peer topologies.

Currently, [MSC4195: MatrixRTC Transport using LiveKit backend](https://github.com/matrix-org/matrix-spec-proposals/pull/4195) 
is the primary proposal, while a full-mesh transport based on 
[MSC3401](https://github.com/matrix-org/matrix-spec-proposals/pull/3401) is also planned. 
Each transport defines its connection model, supported topologies, and any additional requirements 
for participating clients.

#### RTC Transports in `m.rtc.member` Events

Each `m.rtc.member` event contains an `rtc_transports` array, which allows other clients to
determine how to connect to, and exchange real-time data with that participant. This array includes
at least one transport JSON object, each containing at least the following required field:

* `type` (string) — Identifies the RTC transport.  
* Additional fields may be present depending on the RTC transport `type`.

The `versions` field of the `m.rtc.member` event SHOULD be used to determine which transports are

> supported by all participants. Clients SHOULD publish their RTC data using at least one transport

Reviewer: I may have missed something but it's still unclear to me at this point what exactly goes into the `versions` field on member events. Do we define the possible versions and their semantics somewhere?

Reviewer: Clarified out of band, this is intended to provide per-participant capabilities to facilitate migrating between future versions (or versions of the MSC). This will need further thought and description in (or removal from) the proposal.

Reviewer: Our current thinking on this is to drop the `versions` properties on member events. Instead we're considering to add an `encryption` property on the slot event to configure the encryption settings for the session. It's still under discussion whether we'll also fixate the transport to use via the slot.

Reviewer: This has been implemented in the text by now.

MSCs for each transport type.

#### Discovery of RTC Transports

MatrixRTC requires a mechanism for clients to discover which RTC transports — such as an SFU or TURN
server — are available in their Matrix deployment. To support this, homeservers expose an 
authenticated endpoint that returns a list of JSON objects. The endpoint is authenticated as the RTC
transport information is only useful to local clients; remote clients obtain transport information 
via `m.rtc.member` events. Each transport-description object represents an RTC transport as defined
by the corresponding MSC.

> [!NOTE]
> RTC Transports in this context can be anything that can serve as the backend for a MatrixRTC
> session. In most cases this is a SFU. But also a full mesh implementation could be a transport.
> Not all kinds of RTC transports require a way of sourcing a backend resource (e.g. a
> peer-to-peer-solution). In this MSC we only refer to transport where it is necessary to have
> access to additional data to participate in the MatrixRTC session.

The endpoint: `GET /_matrix/client/v1/rtc/transports` is used to expose a sorted (by priority) list

> of Transport description objects.

Reviewer: I vaguely remember that an earlier version of this (or maybe another) proposal used a .well-known file to publish the backends. We should ideally clarify why we're not using .well-known (contrary to e.g. [`/.well-known/matrix/policy_server`](https://spec.matrix.org/v1.18/client-server-api/#getwell-knownmatrixpolicy_server)) in the alternatives section.

Reviewer: Clarified out of band, we switched to a dedicated endpoint because we wanted it to be authenticated and it, therefore, didn't fit well with the `.well-known` concept.

Reviewer: See also https://github.com/element-hq/synapse/pull/18967#issue-3449097319.

Reviewer: Clarified in eac10054304e7251af911ef4d683d8d379274d10.

Response format:

```json5
{
  "rtc_transports": [
    {
      "type": "some-transport-type",
      "additional-type-specific-field": "https://my_transport.domain",
      "another-additional-type-specific-field": ["with", "Array", "type"]
    }
  ]
}
```

Concrete example for a `livekit` transport:

```json5
{
  "rtc_transports": [
    {
      "type": "livekit",
      "livekit_service_url": "https://matrix-rtc.example.com/livekit/jwt"
    }
  ]
}
```

**Fields:**

* `rtc_transports` — array of transport-description representing the available RTC transports
  offered by the homeserver.  
* Each object in the array MUST conform to the JSON schema defined for its `type` (e.g.

>   `livekit` in MSC4195).

Reviewer: Just linkifying for easier navigation.

```suggestion
  `livekit` in [MSC4195](https://github.com/matrix-org/matrix-spec-proposals/pull/4195)).
```

Clients SHOULD use this list to determine which RTC transports to connect to and may advertise their
selected transports according to the respective MSC in the `rtc_transports` field of their
`m.rtc.member` events.

> ### End-to-end Encryption of RTC Data

Reviewer: Discussed out of band with @toger5 and @BillCarsonFr: We could consider breaking this out into a separate proposal to further reduce this proposal's size.

Reviewer: On reconsidering, I think this seems fine for now.

Keys are distributed to the RTC members. The RTC membership is based on top of the room membership
and inherits its security properties.

The key exchange depends on the underlying secure peer-to-peer olm channel (via to-device message)
for key distribution.

This ensures that only members in the RTC slot can decrypt data/media; other people, even if in the
room, will never get the key material.

The RTC transport MSC (e.g., 
[MSC4195: MatrixRTC Transport using LiveKit backend](https://github.com/matrix-org/matrix-spec-proposals/pull/4195)) 
specifies how the key material is actually used. For example, the RTC client may generate a secure
random 256-bit key, which the application can then use to derive the required secrets (using HKDF or
other key-stretching methods as appropriate).

#### Key Distribution

When joining a MatrixRTC slot, each participant shares a generated E2EE key with the other members
(`m.rtc.member`) of the same slot by sending an **encrypted** [Matrix to-device
message](https://spec.matrix.org/v1.11/client-server-api/#send-to-device-messaging). The key is
transmitted via an event of type `m.rtc.encryption_key`. The **target device ID** is taken from the

> `member`.`device_id` field of the recipient’s `m.rtc.member` event. The event follows the following

Reviewer: It looks like this is called `claimed_device_id` now.

```suggestion
transmitted via an event of type `m.rtc.encryption_key`. The **target device ID** is taken from the
`member.claimed_device_id` field of the recipient’s `m.rtc.member` event. The event follows the following
```

schema:

```json5
// event type: "m.rtc.encryption_key"
{
    "room_id": "!roomid:matrix.domain",
    "member_id": "xyzABCDEF0123",
    "media_key": {
      "index": 10,
      "key": "base64encodedkey",
    },
    "version": "0"
}
```

**Field explanations:**

* `member_id` required string: The `member.id` from the target `m.rtc.member` event. Note, because
  `member.id` is globally unique per member instance, it is sufficient to disambiguate multiple key
  events for the same device, even if they use the same `media_key.index` value.  
* `media_key` The media key to use to decrypt  the participant media:  

>   * `key` required string: The base64 encoded key material.  

Reviewer: Is this [unpadded base64](https://spec.matrix.org/v1.16/appendices/#unpadded-base64)?

Reviewer: Clarified out of band, probably yes.

Reviewer: Judging from matrix-js-sdk, it seems to be _padded_: https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/matrixrtc/RTCEncryptionManager.ts#L410

Reviewer: Chatting with @BillCarsonFr and @robintown it looks like we should be using unpadded base64 here. Clarified in e5e135a208b4e5fa60d0b55dbfe1a587ae8fcd4b.

  * `index` required int: The index of the key to distinguish it from other keys. This must be
    between 0 and 255 inclusive. In some implementations of MatrixRTC this may correspond to the
    `keyID` field of the WebRTC [SFrame](https://www.w3.org/TR/webrtc-encoded-transform/#sframe)

>     header.  

Reviewer: If the `index` is incremented on every key rotation and key rotations happen whenever members connect or disconnect, do we not have an overflow issue at `index=255` here in very large and/or very long calls?

Reviewer: Clarified out of band: Yes, but it just rolls around. So there is only an issue here if participants haven't received the key before the wrap-around happens.

* After a short delay `delayBeforeUse,` default: 5 seconds), the sending application switches to the
  new key.  
  * It is possible to overwrite the default delay on a per application basis in case an application
    has specific requirements on security or wants to minimize missed stream data  
  * Also negotiation can be used over data channels to confirm all participants have received the new key.

**Note:** Rotating a key will require to re-send a to-device message to all the participants, and
this is expensive. Clients SHOULD minimize key exchange traffic for rapid joiners/leavers. 

For rapid new joiners: A key rotation grace period (`keyRotationGracePeriod`) is used, if a new
member joins during the grace period, then the same key can be used and shared just to that new
member..

For rapid leavers: The `delayBeforeUse` period is used to coalesce any membership change occurring
in that period and then a single key rotation is scheduled afterward.

`keyRotationGracePeriod` must be greater than `delayBeforeUse` or it will have no effects (default
10s and 5s)

A future version of key exchange could introduce ratcheting, this would reduce the key traffic for

> new joiners (only sends the ratcheted key to the new joiners instead of rotating to all).

Reviewer: I think this belongs into the "Alternatives" section.

Reviewer: Addressed in 0229f899e3e041f5995e61d16da44d226278c99a. I instead formulated this as a potential issue with future resolution.

#### Shared Key alternative

For big calls in a room, it might be interesting to use a shared key system instead of a per-sender
key system.  
In order to use a shared key, a new encrypted room event (`m.rtc.shared_encryption_key`) should be

> sent in the room, and the slot should be updated to include the `event_id` of the shared key event.

Reviewer: This makes me wonder if we should put the entire slot information into an encrypted timeline event and have the slot state event only contain that event's ID? That way we would further reduce the metadata leakage to the home server.

Alternatively, we could also leave the state slot event entirely empty and use the trick from https://github.com/matrix-org/matrix-spec-proposals/pull/4013 to link the timeline slot event to it via a relation. You would only have to change a single event then when modifying a slot and you could still configure power levels per event type to limit slot admin access.

Reviewer: Clarified out of band, this is likely not worth it and we may just argue that state itself will hopefully be encryptable one day.

```json5
// Example: an open slot with shared key encryption
{
  "application": {
    "type": "m.call",
    // optional: app specific slot metadata
    "m.call.id": UUID,    
    "m.shared_key_event": "$000"
  }
}
state_key: "m.call#ROOM" // slot_id
```

The shared key event:

```json5
  "event_id": "$000",
  "origin_server_ts": 1759827668867,
  "type": "m.rtc.shared_encryption_key"
  "content": {
     "key": "base64encodedkey"
   },
```

**Pros**: Scales for big calls as there is no need to distribute n keys via to-device and to rotate
keys on joiner/leavers  
**Cons**: Less security, every room member (and future members) have the key materials even if they
didn’t actively join the call. No automatic key rotation, so if the shared key is compromised past
recording could be decrypted.

Only users with enough power level (moderator, admin) can update the slot in order to enable shared
keys call.

Clients should detect when the slot is configured for a shared key, and then use the shared key
instead of generating and sharing sender keys.

> ### Scope and Responsibilities of MatrixRTC Applications

Reviewer: I think the content of this section should move into the introductory section of the proposal as it helps set expectations.

Reviewer: Addressed in 0a78a66a6ad60ca4c0dc97b591adfd0cae221c8c.

* For the actual session state the **conflict-free** approach is well suited. It should be avoided
  to rollback calls and should not allow a participant to initiate such a rollback and break the
  current ongoing session.  
* For permissions and (shared) metadata, **shared state** is the desired concept. By decoupling the
  concept of a **session** (people communicating in a meeting room with a predefined `slot_id`) and
  the **context** itself (the slot), we can have a conflict-free session concept but a centrally
  managed context with potential state rollbacks that do not necessarily interfere with the session
  itself.

### MatrixRTC Membership Heartbeat Keep Alive

If an RTC Transport is involved like a SFU then per definition the SFU has a very good understanding
of the connectivity of the individual participants. In that case it would be nice to be able to
delegate the ownership of the delayed leave event to that infrastructure component. However,
practical implementation has so far proven that the current implementation of delayed events are
also sufficient in adverse network conditions. So for now it's good enough and considered as an

> optimisation for a future MSC.

Reviewer: This undermines the decision to make the management API for delayed events unauthenticated in https://github.com/matrix-org/matrix-spec-proposals/pull/4140. Delegating management to an SFU seems like the only known use case so far. If this isn't required, we may rather want to keep the API authenticated?

Reviewer: Clarified out of band, this was meant to explain why the delegation is not REQUIRED.

Reviewer: Addressed in 82353657faa632e12b3abc09f8a3340c4fae0104.

* They allow clients to compute the current membership (and session) state reliably without
  contributing to long-term room state bloating.  
* As they are persisted in the DAG like regular events, they can also be used to reconstruct
  previous sessions, supporting session history and retrospective analysis.

Furthermore and In contrast to Matrix room state (as of writing this MSC), sticky events can be
already encrypted improving security significantly. 

### Organising MatrixRTC Members (`m.rtc.member`)as an Array per User

[MSC3401](https://github.com/matrix-org/matrix-spec-proposals/pull/3401) proposed to have one state
event per user with that state event containing an array of RTC memberships.

This introduces two problems:

- Potential inconsistency where one user device overwrites the state of another device during a
  concurrent update.  
- When handling client disconnects, 
- [MSC4140 delayed events](https://github.com/matrix-org/matrix-spec-proposals/pull/4140) **cannot
  reliably maintain an accurate membership state**. This is because, at the time a delayed event is
  scheduled, the current membership array may change before the event is actually emitted, making it

>   impossible to predict the correct state in advance.

Reviewer: I think these bullets may belong together?

```suggestion
- When handling client disconnects, [MSC4140 delayed events](https://github.com/matrix-org/matrix-spec-proposals/pull/4140) **cannot
  reliably maintain an accurate membership state**. This is because, at the time a delayed event is
  scheduled, the current membership array may change before the event is actually emitted, making it
  impossible to predict the correct state in advance.
```

    - What streams to connect to.
    - What data in which format to sent over the RTC channels.

This MSC will focus on the Matrix room state, which can be seen as the most high
level signalling of a call:

## Proposal

Each RTC session is made out of a collection of `m.rtc.member` state events.
Each `m.rtc.member` event defines the application type: `application`
and a `call_id`.
The first element of the state key is the `userId` and the second the `deviceId`.
(see [this proposal for state keys](https://github.com/matrix-org/matrix-spec-proposals/pull/3757#issuecomment-2099010555)
for context about second/first state key.)

### The MatrixRTC room state

Everything required for working MatrixRTC
(current session, sessions history, join/leave events, ...) only
require one event type.

A complete `m.rtc.member` state event looks like this:

```json5
// event type: "m.rtc.member"
// event key: "@user:matrix.domain_DEVICEID"
{
  "application": "m.my_session_type",
  "call_id": "",
  "device_id": "DEVICEID",
  "created_ts": Time | undefined,
  "expires_after": Duration,
  "focus_active": {...FOCUS_A},
  "foci_preferred": [
    {...FOCUS_1},
    {...FOCUS_2}
  ]
}
```

> [!NOTE]  

> > This relies on [MSC3757](https://github.com/matrix-org/matrix-spec-proposals/pull/3757).

Reviewer 6: MSC3757 should be added to the dependencies section at the bottom

for another room member to detect the running session and join it.

We include the device_id in the member content to not rely on the exact format of the state key.
In case [MSC3757](https://github.com/matrix-org/matrix-spec-proposals/pull/3757) is used it would not
be the second element of the state key array.

`created_ts` is an optional property that caches the time of creation. It is not required
for an event that, has not yet been updated, there the `origin_server_ts` is used.

> [!NOTE]
> We introduce `created_ts()` as the notation for `created_ts ?? origin_server_ts`

Once the event gets updated, the origin_server_ts needs to be copied into the `created_ts` field.
An existing `created_ts` field implies that this is a state event updating the current session
and a missing `created_ts` field implies that it is a join state event.
All membership events that belong to one member session can be grouped with the index
`created_ts()`+`device_id`. This is why the `m.rtc.member` events deliberately do NOT include a `membership_id`.

Other then the membership sessions, there is **no event** to represent a rtc session (containing all members).

> Such an event would include shared information, and deciding who has authority over that is not trivial.

Reviewer 6: I continue to trip over whether it is wise to force clients to read all possible `m.rtc.member` events to figure out if a call is happening, and who created it.

I /think/ that a better reason for not having an `m.rtc` state event describing the existence of an RTC session is that you'd have to handle disconnection semantics on it similar to delayed-events for `m.rtc.member`... at which point, why not leverage the membership events?

However, it still feels *REALLY* weird to not have something in state telling you whether semantically a call is intended to be happening now (and what that sort of call is, when it began, and who initiated it) - versus having to infer it.

Reviewer 6: Alternatively, if the thought experiment is "what if two users both create an `m.rtc` state event on different forks of the DAG at the same time?" ... is that really so bad? and does aggregating `m.rtc.member` state actually make it better? if so, how?

In other words, we actually need to justify the lack of `m.rtc` state event much better here, imo.  In particular, having somewhere to store the metadata about the call at the point of creation (its name, its ID, whether it's intended to be a voice/video room or a group call or a conference, etc)

Author: There is a large list of reasons and I 100% support the idea to go into detail in the MSC to justify this approach.

As for this comment I will just give a couple of short examples/arguments:

The security and "trolling" surface is huge. If we have one state event we either limit who can create a call or we allow everyone to mess with the event. This can go from smaller issues like ending the call for fun to larger issues like changing where the call is happening without the members noticing it. (in a LK world at least)

But even if everyone plays fair, a call can be stopped by a delayed event because the creator failed to send the refresh event. This now disconnects everyone from the call.
Independent how we make things behave, if we have a public/shared event controlling the experience for everyone we switch from:
 - If there is a client with issues (or user that actively introduces issues) that user has a degraded experience
 - If there is a client with issues (or user that actively introduces issues) everyone experience can be broken.
 
In the context of matrix where there is no central entity controlling the clients the seconds seems to be the only valid option.

Author: The slot state events seem to be a great balance between the two positions brought up in this thread.

Do we feel comfortable closing this thread?

Reviewer: I think we can consider this resolved given that we now have a good compromise of slots in state vs. member events in the room.

correctly representing the current RTC session state. Historic and current MatrixRTC session data would be broken.

For an acceptable solution, the following requirements need to be taken into consideration:

- Room state is set to empty if the client loses connection. (A heardbeat like system is desired)
- The best source of truth for a call participation is a working connection to the SFU.
  It is desired that the disconnect of the SFU is connected to the room state.
- It should be possible to update the room state without the client being online.
- All of this should still work when Matrix uses cryptographic identities (e.g.
  [MSC4080](https://github.com/matrix-org/matrix-spec-proposals/pull/4080)).

[MSC4140](https://github.com/matrix-org/matrix-spec-proposals/pull/4140) proposes a concept to
delay the leave events until one of the leave conditions (heartbeat or SFU disconnect) occur
and fulfil all of the these requirements.

A MatrixRTC client has to first send/schedule the following delayed leave event:

```json5
// event type: "m.rtc.member"
// event key: "@user:matrix.domain_DEVICEID"
{

>   "leave_reason": "CONNECTION_LOST"

Reviewer 6: this should be namespaced. what other leave reasons are there?

Reviewer: This seems to refer to a previous version of the error schema. The new schema has separate categories so I'm considering this resolved.

  - For custom deployments that benefit from having the Focus configuration on a per client basis instead of per homeserver.

### The RTC Session types (application)

Each session type can have its own specification in how the different streams
are interpreted and even what focus type to use. This makes this proposal extremely
flexible. For instance, a Jitsi conference could be added by introducing a new `application`
and a new focus type and would be MatrixRTC compatible. It would not be compatible
with applications that do not use the Jitsi focus but clients would know that there
is an ongoing session of unknown type and unknown focus and could display/represent
this in the user interface.

To make it easy for clients to support different RTC session types, the recommended
approach is to provide a Matrix widget for each session type, so that client developers
can use the widget as the first implementation if they want to support this RTC
session type.

Each application should get its own MSC in which the all the additional
fields are explained and how the communication with the possible foci is
defined:

- [`m.call`](www.example.com) TODO: create `m.call` MSC and add link here.

> ## Potential issues

Reviewer 6: The fact we don't reference how to tell users that a call is happening (i.e. [m.call.notify](https://github.com/matrix-org/matrix-spec-proposals/pull/4075)) is very disorienting here.

Reviewer: The latest version of https://github.com/matrix-org/matrix-spec-proposals/pull/4075 introduces generic MatrixRTC notifications now. So we could at least link to it from here.

Reviewer: I have added notifications and #4075 as a separate bullet in the intro with 6bf733fb2773ba9de1f7323dd2942a15cabbc5c6.

> - [`m.call`](www.example.com) TODO: create `m.call` MSC and add link here.

Reviewer 6: Something which is very unclear right now is the lifecycle of a call (and the draft of MSC4196 doesn't really help).

Specifically, what has happened to `m.call.invite`? `m.call.answer`? `m.call.select_answer`? `m.call.hangup`? The intention of MSC3401 was to keep using these (as per https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/group-voip/proposals/3401-group-voip.md#basic-call) even when an SFU is involved, in order to preserve signalling for:

 * Call was placed but isn't ringing yet (not that we have in legacy VoIP)
 * Call is ringing
 * Caller gave up and stopped ringing
 * Call has been answered on a device
 * Call was rejected locally
 * Call was rejected on all devices
 * Call was answered
 * Call was hung up

i.e. all the state machine transitions that go into the lifecycle of a call.

I'm very worried that this seems to have been collapsed to a single `m.call.notify` event from MSC4075 + `m.rtc.member` events here, which seems to be missing most of the above. For instance, I can't see how the caller tells the callee that it's no longer ringing them - not to mention stuff like early media (MSC3635) or ringback tones.

Legacy VoIP was already a dumbed down version of SIP, and I'm worried that the lack of signalling semantics here is going to make bridging to SIP or the PSTN really hard, and a backwards step from legacy voip.

Author: The reason that information about this is so thin here is that this proposal tries to define the line between matrixRTC and the special case matrix rtc use case of calls.

The focus should be to find the minimum requirements for a rtc session so clients can implement everything from this msc and they will be able to have a rough idea what is happening in a room.

Everything that is not applicable to all kinds of sessions should be part of their dedicated msc's.
Calling in particular has a very rich setup/notification cycle that is not necessarily part of every rtc session.

Th idea very much is however to fully support this list:

-     Call was placed but isn't ringing yet (not that we have in legacy VoIP)
-     Call is ringing
-     Caller gave up and stopped ringing
-     Call has been answered on a device
-     Call was rejected locally
-     Call was rejected on all devices
-     Call was answered
-     Call was hung up

This proposals tries to not enforce this for matrixRTC in general however.

Your comment makes me wonder if there is any benefit in reusing the already existing legacy calling events for `m.call.invite` `m.call.answer` `m.call.select_answer` `m.call.hangup` ...

Reviewer 6: > this proposal tries to define the line between matrixRTC and the special case matrix rtc use case of calls.

are we *sure* this is not an unnecessary abstraction right now? we could add it later, if/when we have a use case for MatrixRTC which doesn't involve calling? and meanwhile keep the MSCs easier to follow and less fragmented?

Reviewer 6: > Your comment makes me wonder if there is any benefit in reusing the already existing legacy calling events for m.call.invite m.call.answer m.call.select_answer m.call.hangup ...

There may not be any advantage to reusing those events, but it's very hard to tell when I can't figure out how those semantics map onto a MatrixRTC call life cycle today.  My spidey sense is that having to infer the call state machine out of m.rtc.member and m.rtc.notify events is going to be fragile, hard to reason about, and hard to extend (e.g. how would you do early media, for compatibility with MSC3635?) - rather than throwing explicit events around when things happen.  But am very happy to be corrected... especially if the rationale ends up in an MSC :)

Reviewer: I have moved this to #4075 in https://github.com/matrix-org/matrix-spec-proposals/pull/4075/changes#r3427055308 as it pertains to notifications / calls.

## Alternatives

## Security considerations

## Unstable prefix

The state events and the well_known key introduced in this MSC use the unstable prefix

> `org.matrix.msc4143.` instead of `m.` as used in the text.

Reviewer 6: empirically we seem to be using `org.matrix.msc3401.call.member` rather than `org.matrix.msc4143.rtc.member` in Element?

Author: The current implementation is still using the msc3401 prefix. That is wrong and will be addressed. There is still the open topic of how exactly we want to do the state keys  and the event ownership and on top of that we have plans for how to index rtc member events in a better way.
The reason we changed it to (`...call...` ->) `...rtc...` is, that we need the `call` namespace for the particular video call matrixRTC application (session type) of calling over MatrixRTC. Using that word for both. Matrix rtc sessions in general and calls will be confusing in the long run.

Reviewer: The switch to `rtc` seems sane to me given that calls our now a special RTC application type.

    - What MatrixRTC backends are supported.
- End-to-end encryption of media streams

This MSC will focus on the Matrix room state which is responsible for the high
level signalling of a RTC session:

## Proposal

Each RTC session is made out of a collection of `m.rtc.member` room state events.
Each `m.rtc.member` event defines who (the `member`) is a participant of which session (the `session`).

### The MatrixRTC room state

All data related to a MatrixRTC session
(current session, sessions history, join/leave events, ...) only
requires one event type.

(current session, sessions history, join/leave events, ...) only
require one event type:.

We use a set of `m.rtc.member` (one for each participant) state events to represent a session.

based on the content a `m.rtc.member` state event can either represent a connected or a disconnected member.

#### Joining a session

Sending a well-formed `m.rtc.member` event that describes a connected state for a state key that is not yet used or contains a disconnected `m.rtc.member` event represents a join action. 

The fields are as follows:

- `member` required object - describes the participant of the RTC session:

>   - `id` required string - a unique identifier for this session membership as defined above. Recommended to be a UUID. It can be reused if the user leaves and rejoins the session.

Reviewer: Maybe obvious but presumably the recommendation is for UUIDv4?

Reviewer: This is obsolete given that we don't mention UUIDs explicitly for `member.id` anymore.

  },
  "member": {
    "id": "xyzABCDEF10123",
    "device_id": "DEVICEID",
    "user_id": "@user:matrix.domain"
  },
  "created_ts": Time | undefined,
  "focus_active": {...FOCUS_A},
  "foci_preferred": [
    {...FOCUS_1},
    {...FOCUS_2}
  ]
}
```

This gives us the information, that user: `@user:matrix.domain` with member ID `DEVICEID_m:call_123456789`
is part of a session identified by `{}` using application of type `m.call` connected over `FOCUS_A`.
This is sufficient information for another room member to detect the running session and join it.

`created_ts` is an optional property that caches the time of creation. It is not required
for an event that, has not yet been updated, there the `origin_server_ts` is used.

> [!NOTE]
> We introduce `created_ts()` as the notation for `created_ts ?? origin_server_ts`

Once the event gets updated the `origin_server_ts` needs to be copied into the `created_ts` field.
An existing `created_ts` field implies that this is a state event updating the current session
and a missing `created_ts` field implies that it is a join state event.
All membership events that belong to one member session can be grouped with the index

> `created_ts()`+`state_key`. This is why the `m.rtc.member` events deliberately do NOT include something akin to a `membership_id`.

Reviewer: I'm curious why the chain is being build via the timestamp of the initial event and not its ID?

Author: The timestamp is needed anyhow to compute expired events. So this is one field that is the same in all events and where we do not need an iterative approach. In the end this is an implementation detail. The spec only clarifies that this is a required property (one an implementation is allowed to rely on) of a membership.

```
membershipsForSingleDevice.map((m)=>m.created_ts ?? m.origin_server_ts).filter((ts)=>ts===related_events_ts)
```
Is a valid implementation to collect all events for connected session.

It still does not help with finding the associated leave event.
It is still helpful in making clear how `created_ts` behaves.

> Other then the membership sessions, there is **no event** to represent a RTC session (containing all members).

Reviewer: ```suggestion
Other than the membership sessions, there is **no event** to represent a RTC session (containing all members).
```

This event would include shared information where it is not trivial to decide who has authority over it.
Instead the session is a computed value based on `m.rtc.member` events.
The list of events with the same `session` content represent one session.
This array allows to compute fields like participant count, start time etc.

Based on the value of `application`, the event might include additional parameters
to provide additional session parameters.

> A [Third Room](https://thirdroom.io) like experience could include the information of an approximate position
> on the map, so that clients can omit connecting to participants that are not in their
> area of interest.

> #### State key for `m.rtc.member`

Reviewer 7: Several applications (widgets) can send this event to the same room. In most cases the application should only receive and send it's own member events, so that data is not shared and cannot be modified by another application. Unless someone writes a widget to show and or manage all rtc data in the room. 

It would be nice if the spec could define this behavior. Maybe we could have some changes to widget capabilities or some default behavior for this state event.

Author: Based on what field would the widget api/driver filter the events. They all have the same type.

Reviewer: Using MatrixRTC in widgets has been descoped from this proposal. It can be layered on top of this MSC in a separate proposal (but will obviously block on the entirety of widget spec also being in MSC stage).

The state key is generated from the `member` field of the `m.rtc.member` event.

We want to choose a state key that is compatible with whichever state protection proposal is accepted to ensure that
users cannot modify one another's sessions.

For [MSC3757](https://github.com/matrix-org/matrix-spec-proposals/pull/3757) we generate the state key by
concatenating the following strings:

- the Matrix ID of the user
- an `_` (underscore)

> - the `member`.`id` field

Reviewer 7: Several applications can send this `m.rtc.member` event. There are existing ideas to use `application` as part of the state key, using extra `_`, or even hash `device_id`, `application`. In this case each application will have members data separated. It would be nice to have it defined in the spec.

Author: This has been added

For example with a `member`.`id` of `xyzABCDEF10123` for user `@user:matrix.domain` the state key would be `@user:matrix.domain_xyzABCDEF10123`.

For a client parsing the state key we would treat anything before the first `_` as the Matrix ID of the user
and anything after as the `member`.`id` field.

#### Leaving a session

> Sending an empty `m.rtc.member` event represents a leave action. The state key must be the same as boefore

Reviewer: It is said further up that the session participant (`member.user_id`) is not necessarily identical to the sender of the `m.rtc.member` event. How does this interact with state protection when a user wants to leave a session that he was added to by somebody else?

Author: This is a really good reason, that the state key should not be used as the state protection. We would like the external user to create a session for you and allow you + them to modify this state key.

  - `index` required int: The index of the key to distinguish it from other keys. This must be a between 0 and 255 inclusive.
    In some implementations of MatrixRTC this may correspond to the `keyID` field of the WebRTC [SFrame](https://www.w3.org/TR/webrtc-encoded-transform/#sframe) header.
  - `invalidates_key_index` optional int: The index of the key that is invalidated by this key. If this is set, the application should invalidate the key identified
    by `invalidates_key_index` once it receives a frame with the new `index`. This is to protect against an exfiltrated key being used to forge frames.
  - `invalidates_after_ms` optional int: The number of milliseconds after the key identified by `invalidates_key_index` is invalidated by this key even if no frames
    are received. Again, this is to protect against an exfiltrated key being used to forge frames.

Depending on the RTC application, additional fields may be added to this event.

An example to-device event:

```json5
// event type: "m.rtc.encryption_keys"
{
    "session": {
      "application": "m.call",
      "call_id": "",
      "scope": "m.room"
    },
    "member": {
      "id": "xyzABCDEF10123",

>       "device_id": "DEVICEID",

Author: There is no olm session for the `m.call.member` state events.
And we might want the option to create memberships from other accounts.

>       "device_id": "DEVICEID",

Reviewer: There's also no Olm session in unencrypted rooms. So I think we'll need these. Considering this resolved given time and the renaming.

to provide additional session parameters.

> A [Third Room](https://thirdroom.io) like experience could include the information of an approximate position
> on the map, so that clients can omit connecting to participants that are not in their
> area of interest.

#### State key for `m.rtc.member`

The state key is generated from the `member` field of the `m.rtc.member` event.

We want to choose a state key that is compatible with whichever state protection proposal is accepted to ensure that
users cannot modify one another's sessions.

For [MSC3757](https://github.com/matrix-org/matrix-spec-proposals/pull/3757) we generate the state key by
concatenating the following strings:

- the Matrix ID of the user
- an `_` (underscore)
- the `member`.`id` field

For example with a `member`.`id` of `xyzABCDEF10123` for user `@user:matrix.domain` the state key would be `@user:matrix.domain_xyzABCDEF10123`.

For a client parsing the state key we would treat anything before the first `_` as the Matrix ID of the user
and anything after as the `member`.`id` field.

The state key should/can **never** be used to imply any information about the user device or application.

> Even thought the proposed format is`<device_id>_<application><optional_application_id>`,
> the state key only has to fulfill be unique in regards to device, application and application_id.

Reviewer 8: ```suggestion
Even though the proposed format is`<device_id>_<application><optional_application_id>`,
the state key only has to be unique in regards to device, application and application_id.
```
This makes it grammatically correct, but I still struggle to understand it. Do you mean that the state key is a unique combination from these 3 variables?

Reviewer: This is obsolete by now.

members from decrypting past/future RTC data.

Additionally, clients SHOULD also rotate their key on a periodic schedule regardless of whether
the members have changed. This limits the impact of compromised keys.

In order to account for the delivery latency of to-device messages, clients SHOULD add a short
delay after sending a new key before starting to use it. Otherwise, receiving members may
be unable to decrypt the sender's streams temporarily. The RECOMMENDED delay duration is 5 seconds.

Furthermore, resending to-device messages to all members can be expensive when multiple
members join and/or leave in short succession. To mitigate this, clients MAY apply
some flexibility to exactly when a rotation happens relative to a membership change. This means
accepting a small window in which joining or leaving members could decrypt media that is
slightly outside their actual membership period in exchange for fewer key rotations.

As an example, let's assume a client applies the time interval `delay = 5s` between rotating a key
and starting to use it for encryption. When members leave during `delay`, the client

> could schedule another rotation for when `delay` has elapsed. This coalesces multiple member

Reviewer: To address https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3852135729.

```suggestion
and starting to use it for encryption. When any members leave in the `delay` time window
following a key rotation, the client could collect all leavers within that window and then
schedule another rotation once `delay` has elapsed. This coalesces multiple member
```

objects are not required on closed slots but may be kept around for convenience to simplify re-opening
the slot. The semantics of open and closed slots for actual slot membership are described in the membership
event section [below].

Slots do not close automatically. A slot that is deliberately long-lived could be used to
create a Discord-style experience where members hop on and off as desired. A conferencing
or meeting-centric client might close its slots at the end of the meeting. A future
proposal might change how slots are automatically managed.

### Membership

Membership in slots is expressed via `m.rtc.member` room events. These events provide sufficient
metadata for other room members to join the same slot and to exchange media streams via the
chosen transports.

`m.rtc.member` events MUST be sent as sticky events as per [MSC4354: Sticky Events][MSC4354]. This
results in the same delivery guarantee that state events have which is highly desirable for RTC
experiences. At the same time, it avoids the drawbacks associated with state events. Further details
on this can be found in [MSC4354]. Clients MUST also implement the ephemeral map algorithm as defined

> in the addendum of [MSC4354] to construct a state-like store of membership events.

Reviewer 9: Do we want to say what sticky duration should be used?

Reviewer: Good point. Our implementations use 1 hour which seems like a sane recommendation to avoid having to refresh it too often while still being effective. Have added that with 59c3f98c424e73629bb56587de49cafbe702afae.

[MSC4354]: https://github.com/matrix-org/matrix-spec-proposals/pull/4354

Within `m.rtc.member` events, `content` contains the following properties:

- `slot_id` (required, string): The `state_key` of the slot that is being joined.
- `member` (required, object): Information to identify the member.
  - `id` (required, string): Identifier to distinguish multiple members, even for the same user
    and device. MUST be unique for each join of the same user. This means that clients need to use a different

>     identifier when leaving and then rejoining a slot.

Reviewer 9: ```suggestion
    identifier when leaving and then rejoining a slot, and must be unique across different devices.
```

Given sticky events are keyed per-user, we need the `member.id` to be unique across different devices, i.e. use a random string generator rather than a counter.

Assuming we want to support multiple devices? (I assume so?)

    allowing clients to avoid connecting to members outside their area of interest.
- `transports` (object): Details on the MatrixRTC transports of this member. Other clients use the
  information in this object to determine how to connect to and exchange real-time data with this
  member. Clients should be prepared to connect to as many transports as there are members
  joined to the session. The exact procedure for publishing and subscribing to real-time data is
  defined in each transport's specification.
  - `published` (array): An array of objects describing the transports on which the member is
    publishing media.
    - `type`: (required, string): The [MSC4519]-registered transport `type`. For an [MSC4195]
      transport, this would be `livekit`.
    - Optionally includes further properties specific to the transport `type`. The concrete properties
      are defined by the transport's specification. This could, for instance, include WebSocket URLs.
      The transport's specification would be as per its [MSC4519] registration.
  - `can_subscribe` (array): An array of transport types that the member is able to subscribe to.

>     Other members can use this as cue for deciding which transports to use to accommodate this member.

Reviewer 9: Do we also need a section for how `transports` work in the abstract? Like how does a client choose which one to use? What happens if there is no overlap of supported transport types?

Also it says `can_subscribe` can be used to decide what transports to start to accommodate other members? But wouldn't you want to have every type you support published anyway? Or does everyone join initially with empty `published` and then decides what to publish?

Reviewer: > Do we also need a section for how `transports` work in the abstract? Like how does a client choose which one to use? What happens if there is no overlap of supported transport types?

Yes, good point. Our thinking was to offload any negotiation or consensus for choosing transports to a future MSC. Right now we only have one transport (LiveKit). How exactly clients agree on which transport to choose depends on what transports are available. So I think we cannot prescribe that generically. The future MSC that introduces the second transport will need to cover how clients choose between LiveKit and the new transport and what happens if there is no overlap in client support.

> But wouldn't you want to have every type you support published anyway? Or does everyone join initially with empty `published` and then decides what to publish?

No, I think you wouldn't. Publishing a transport will usually incur some amount of cost (traffic, processing, etc.) on clients. So you clients have an interest to only publish transports that are actually needed.

Have tried to clarify both in d6133e48cb6876f2e722f101a34da8c42877f142.

  the `m.rtc.member` event's `slot_id`.
- The sender is currently a member of the room (i.e. has room membership `join`).
- The event is currently sticky, meaning that its stickiness duration as per [MSC4354] has not expired.
  This is to ensure that the membership view is as consistent as possible across all members.

If these conditions are not fulfilled, clients MUST treat the member as left and refrain
from connecting to their transports.

#### Leaving a slot

To voluntarily leave a slot, the client sends an `m.rtc.member` event for the desired `member.id` and with `membership = leave`.
`m.rtc.member` event.

```json5
{
  "type": "m.rtc.member",
  "content": {
    "slot_id": "{application_type}#{application_slot_id}", // = m.rtc.slot state_key
    "member": {
      "id": "{member_id}",
      "membership": "leave",
    },
    "leave_reason": {
      "code": "{code}",
      "reason": "{reason}",
    },
    "sticky_key": "{member_id}" // = member.id from previously joined m.rtc.member event
  },
  ...
}
```

Again, once a member has left, clients SHOULD refrain from connecting to their transports.

[lifecycle]: #membership-lifecycle

> #### Membership lifecycle

Reviewer 9: Might be worth noting what happens with redactions? Presumably the're treated as leaves? So a mod can kick users off a call?

Reviewer: Yes, exactly. Have clarified it with 0e473cb79b7bfaea78b994e523f8cde7a0e04060.

data itself. This process is generally specific to the transport being used, but often requires session
members to agree on key material, at a minimum. To support this, MatrixRTC provides a generic system
for establishing shared key material between members. Transports can then define how to actually use
this key material, which may involve deriving further secrets from it. The concrete mechanism for
agreeing on the shared key material within a slot is prescribed through the `encryption` object in
`m.rtc.slot` events.

[existing mechanisms]: https://spec.matrix.org/v1.19/client-server-api/#end-to-end-encryption

Use of encryption in MatrixRTC is REQUIRED in encrypted rooms. This means that `m.rtc.member` events
MUST be encrypted and `m.rtc.slot` events MUST contain an `encryption` object when sent in an encrypted
room. Member / slot events that violate these conditions MUST be considered left / closed.

The only available encryption mechanism for now is `m.per_member` which is defined in the remainder of
this section.

```json5
{
  "type": "m.rtc.slot",
  "content": {
    "encryption": {
      "type": "m.per_member"
    },
    ...
  },
  ...
}
```

> Under `m.per_member` every member maintains a unique sender key. This key is shared securely

Reviewer 10: How is the key used to encrypt?  What algorithms are used?  Should the algorithm names be part of the `type`?  Is the method transport-specific?

with other members via Olm-encrypted [to-device messages]. This ensures that keys are only
distributed among session members. Other devices, even if in the room, never get the key material.

[to-device messages]: https://spec.matrix.org/v1.18/client-server-api/#send-to-device-messaging

#### Distributing keys

When joining a slot, clients generate a 32-byte key by using a cryptographically secure pseudorandom
number generator. They then share the key with other clients joined to the slot by sending encrypted

> to-device messages of the type `m.rtc.encryption_key`.

Reviewer 9: Do we need to worry about missed `to-device` messages? I think megolm allows rerequesting keys in case of wedged OLM session?

Reviewer 11: There is no re-requesting of RTC keys, and for now no plan to support it in the future.
This behavior in megolm is also pretty much deprecated.

The recipient devices are determined from the `m.rtc.member` events that are considered to be
joined to the slot. The conditions for considering a member joined were given
[above](#joining-a-slot). Once the member events are determined, the `m.rtc.encryption_key`

> to-device messages are sent to the devices that were used to encrypt these member events.

Reviewer 9: Is this obvious how to do correctly? Given that e.g. `device_id` and `sender_key` are untrusted and shouldn't really be used? Might want to be explicit here if this is a non-standard thing to do

The schema for `m.rtc.encryption_key` to-device messages is as follows:

```json5
// PUT /_matrix/client/v3/sendToDevice/m.rtc.encryption_key/{txnId}
// Unencrypted content of OlmPayload shown, but in reality this would be an encrypted message

{
  "room_id": "{room_id}",
  "member_id": "{member_id}",
  "media_key": {
    "index": <index>,
    "key": "{encoded_key}",
    "format": "m.base64"
  }
}
```

- `room_id` (required, string): The ID of the room that the slot is located in.
- `member_id` (required, string): The `member.id` value of the sender's `m.rtc.member` event.
  Note that because `member.id` is unique per member, it is sufficient to disambiguate multiple
  key events for the same device.
- `media_key` (required, object): Information on the key material.
  - `key` (required, string): The key (32 bytes) encoded as specified by `format`.
  - `index` (required, number): The rolling index of the key to distinguish it from other keys. The
    value MUST be between 0 and 255 inclusive. WebRTC-based transports may use this as the `keyID`

>     field of [SFrame](https://www.w3.org/TR/webrtc-encoded-transform/#sframe) headers.

Reviewer 9: What happens if index reaches 255?

Reviewer 11: It should wrap around 0

the value of `member_id` in the `m.rtc.encryption_key` message. Once the member event was determined,
clients verify that the sender and device that was used to send the member event match the sender
and device of the to-device message. Otherwise the message MUST be discarded.

In keeping with [MSC4153: Exclude non-cross-signed devices][MSC4153], clients SHOULD also discard
`m.rtc.encryption_key` events when the sending device is not cross-signed by its owner.

[MSC4153]: https://github.com/matrix-org/matrix-spec-proposals/pull/4153

#### Rotating keys

To ensure confidentiality, clients SHOULD rotate and redistribute their key whenever the set of
members that are considered joined to the slot changes. Rotation prevents joining/leaving
members from decrypting past/future RTC data.

Additionally, clients SHOULD also rotate their key on a periodic schedule regardless of whether
the members have changed. This limits the impact of compromised keys.

In order to account for the delivery latency of to-device messages, clients SHOULD add a short
delay after sending a new key before starting to use it. Otherwise, receiving members may

> be unable to decrypt the sender's streams temporarily. The RECOMMENDED delay duration is 5 seconds.

Reviewer 9: The reverse race also exists, the `to_device` message might arrive before the member event. Probably worth noting that clients should also store `to_device` messages for unknown members for N seconds

Furthermore, resending to-device messages to all members can be expensive when multiple
members join and/or leave in short succession. To mitigate this, clients MAY apply
some flexibility to exactly when a rotation happens relative to a membership change. This means
accepting a small window in which joining or leaving members could decrypt media that is

> slightly outside their actual membership period in exchange for fewer key rotations.

Reviewer 9: We skip straight to reducing number of rotations without actually specifying what happens when someone joins the call? Presumably everyone immediately sends the keys to the new joiner? Do they rotate as well?

Reviewer: This is covered at the very beginning of this section. Whenever somebody joins (or leaves) keys should be rotated and sent to everyone (including the new joiner).

> #### Rotating keys
>
> To ensure confidentiality, clients SHOULD rotate and redistribute their key whenever the set of
> members that are considered joined to the slot changes.

This feels unambiguous enough to me but I'm happy to amend it in case you have suggestions.

# MSC4143: MatrixRTC – Real-time communication over Matrix

Matrix is a generalised protocol for decentralised communication. This includes chatting but also
real-time communication (RTC) such as VoIP. While Matrix supports VoIP signalling for [1-to-1 calls],
and [MSC3401] attempts to extend it to group calling, a unified system for RTC applications is
currently missing.

[1-to-1 calls]: https://spec.matrix.org/v1.18/client-server-api/#voice-over-ip
[MSC3401]: https://github.com/matrix-org/matrix-spec-proposals/pull/3401

The present proposal aims to close this gap by introducing *MatrixRTC*, a generalised framework for
building RTC experiences on top of Matrix. At a high level, MatrixRTC consists of the following parts:

* **End-to-end encryption** provides a generic basis for encrypted media exchange and reuses existing
  Matrix primitives such as encrypted room and to-device messages.
* **Transports** define how members exchange media streams. This can, for instance, happen
  peer-to-peer or through Selective Forwarding Unit (SFUs). Transports also determine how the generic
  end-to-end encryption is used in transport-specific encryption. Transports are recorded in a registry

>   defined by [MSC4519].

Reviewer: This sounds very similar to https://github.com/matrix-org/matrix-spec-proposals/pull/4143#discussion_r3855501758. I'm going to resolve this comment so that we can discuss the question in the other thread.

# MSC4143: MatrixRTC – Real-time communication over Matrix

Matrix is a generalised protocol for decentralised communication. This includes chatting but also

> real-time communication (RTC) such as VoIP. While Matrix supports VoIP signalling for [1-to-1 calls],

Reviewer 12: I've asked privately to folks and got confirmation that the intention is for MatrixRTC to fully replace the current VoIP signalling; this is _not_ clear to me from this MSC.

It is fine if this MSC is not intending to deprecate it yet, but I _think_ it really is? If not, we should at least mention / have an MSC that _does_ deprecate it.

Reviewer: I think this MSC alone cannot do it as it only provides the infrastructure needed to build arbitrary RTC applications. Combined with the other MSCs, the MatrixRTC system should certainly supersede legacy VoIP though, you're right. I have opened https://github.com/matrix-org/matrix-spec-proposals/pull/4531 for doing the formal deprecation and referenced it in this proposal with 2a45b770eb9b26e80431743bbd22bf7043eaf06e.

    [Common Namespaced Identifier Grammar]. MUST align with the event's `state_key`.
  - Optionally includes further properties for settings that are specific to the application
    `type`. The concrete properties are defined by the application's specification. A calling
    application, for instance, could include properties for constraining the call to be voice-only.
- `encryption` (object): If present, describes the encryption mechanism to use in this slot. Further
  details on the available mechanisms can be found in the [encryption section] below. If absent,
  encryption is disabled.
  - `type` (required, string): The globally unique identifier of the encryption mechanism.
    MUST follow the [Common Namespaced Identifier Grammar].
  - Optionally includes further properties for settings that are specific to the encryption
  `type`.

[encryption section]: #end-to-end-encryption

#### Slot lifecycle

A slot is opened by sending an `m.rtc.slot` state event with `status = "open"`, a valid application
object and, if needed, a valid encryption object as per the JSON schema above. Any slot that

> doesn't fulfill these requirements is closed. To explicitly close an open slot, the associated `m.rtc.slot`

Reviewer 13: I would say, "should (or perhaps MUST) be treayed as closed" otherwise it should like some omnipotent being is going around actively closing slots that don't comply.

Reviewer: Have tried to clarify this with 39c144dc092d0e12f2a1b8de445cad9297ef55f9.

for encrypting messages in rooms. On the other hand, applications also need a way to encrypt the RTC
data itself. This process is generally specific to the transport being used, but often requires session
members to agree on key material, at a minimum. To support this, MatrixRTC provides a generic system
for establishing shared key material between members. Transports can then define how to actually use
this key material, which may involve deriving further secrets from it. The concrete mechanism for
agreeing on the shared key material within a slot is prescribed through the `encryption` object in
`m.rtc.slot` events.

[existing mechanisms]: https://spec.matrix.org/v1.19/client-server-api/#end-to-end-encryption

Use of encryption in MatrixRTC is REQUIRED in encrypted rooms. This means that `m.rtc.member` events
MUST be encrypted and `m.rtc.slot` events MUST contain an `encryption` object when sent in an encrypted
room. Member / slot events that violate these conditions MUST be considered left / closed.

Conversely, MatrixRTC encryption MUST NOT be used in unencrypted rooms. This is because the specific

> encryption mechanism introduced in this proposal is not well suited for unencrypted rooms. A future MSC

Reviewer 13: Would be good to explain why

Reviewer: Yes, agreed. Have tried to make it more clear in 98e5783f6b84cdde4905abf5857278edb6f29d11.

may introduce another mechanism that lends itself better to unencrypted rooms.

> The only available encryption mechanism for now is `m.per_member`.

Reviewer 13: Are we defining this here? And is this an encryption mechanism or a key distribution mechanism?

Reviewer 11: Not sure what is the best way to name that. There is another mecanism, names `m.shared` that will be part of another MSC, with that mecanism everyone use the same key to encrypt/decrypt, the plan is to share this key via a room message (better fit for big calls). So it is in fact a different distribution mecanism (room message instead of to-device messages), but it also possible that the shared key would be inside the slot state event (for clear rooms).

We want to tell the difference because `per_member` has more security features than `shared`

Reviewer: > Are we defining this here?

Yes, the whole remainder of this section deals with `m.per_member`. I tried to make this more clear with cea50143e0068efcd5a323b5cd00f60c6de4dc02.

> And is this an encryption mechanism or a key distribution mechanism?

I guess it is more of a key distribution and rotation mechanism, yes. But that felt too much of a mouthful. Also, the introductory paragraph of this section should make it quite clear what the "encryption system" in MatrixRTC is:

> [...] MatrixRTC provides a generic system
> for establishing shared key material between members. Transports can then define how to actually use
> this key material, which may involve deriving further secrets from it. The concrete mechanism for
> agreeing on the shared key material within a slot is prescribed through the `encryption` object in
> `m.rtc.slot` events.

We're happy to take suggestions for rephrasing this to improve readability though.

```json5
{
  "type": "m.rtc.slot",
  "content": {
    "encryption": {
      "type": "m.per_member"
    },
    ...
  },
  ...
}
```

Under `m.per_member` every member maintains a unique sender key. This key is shared securely
with other members via Olm-encrypted [to-device messages]. This ensures that keys are only
distributed among session members. Other devices, even if in the room, never get the key material.

[to-device messages]: https://spec.matrix.org/v1.18/client-server-api/#send-to-device-messaging

#### Distributing keys

> When joining a slot, clients generate a 32-byte cryptographically secure key. They then share

Reviewer 13: Bit weird that we're defining the key to be 32 bytes and "cryptographically secure" without defining what the encryption actually is?

Reviewer 11: The RTC layer by itself does not do any encryption and it is not aware of the type of encryption that the app will do, nonetheless it allows to distribute some "source key material" in the form of a secure random 32-bytes vector. 
The application can then use this input key material to derive/extract a key that fits their use case. For example livekit will use HKDF with a salt `LKFrameEncryptionKey` to derive the AES key to encrypt frame.

Some other apps may need more bytes, like an authentication and an encryption key, they can also use this input key material and expand it to get the correct format they need.

Reviewer 14: I'd agree that "cryptographically secure" is a bit weird here – what does it mean for some bytes to be "secure"?
Is the requirement actually that the bytes are random? Then I would advise to say something like "generated using a CSPRNG (Cryptographically Secure Pseudorandom Number Generator)", or at least "cryptographically _random_" instead of _secure_. (Also see [this](https://github.com/matrix-org/matrix-spec-proposals/pull/4140#discussion_r3529519943) similar discussion on the delayed events MSC that still contained delegation at that point.)
Is the requirement something different? Then I think that should be specified.

Reviewer: Yeah, I think this is the same requirement as on the other proposal. We only care about the bytes being random enough (so that they are actually fit for being used as an encryption key). Have clarified this with e2a64eb989a3bc98c44da4a8bd01337b1742d96c.

Reviewer: > [...] although I could still use a bit more explanation on why the RTC layer has to generate a fixed bit of entropy and the crypto layer has to like it rather than letting the crypto layer generate its own keys and then, in the spec, this is just a string with presumably a sensible max length. Aren't the RTC layer and the crypto layer both just different bits of the client anyway?

You can publish more than one transport simultaneously. So if we let transports define the shared key material, then `media_key` in the to-device message would have to become a mapping of transport type to key. We could do that but I wonder if that's really significantly different to what the proposal currently does.

The idea is that the shared secret generated here has sufficient length / entropy to be used by any transport. If we let each transport generate its own shared secret and put it into a max-lengthed string field, I think the only real difference is that some transports might use the entire max length available in the string field while others use shorter keys?

Reviewer 13: Fair enough. While having transports use keys of length appropriate to their encryption sounds sensible, having to have multiple keys for different transports sounds like a pain. This feels like useful info to remember why the decision was made though, so if it could go into the MSC non-normatively, that would be useful.

Reviewer: Yes, agreed. Have added more explanation in c3d05cfa3cb8ad85262d53c9aea9cff71a04fc4e and also removed the "pseudo" prefix to RNG.

the key with other clients joined to the slot by sending encrypted to-device messages of the
type `m.rtc.encryption_key`.

The recipient devices are determined from the `m.rtc.member` events that are considered to be
joined to the slot. The conditions for considering a member joined were given
[above](#joining-a-slot). Once the member events are determined, the `m.rtc.encryption_key`
to-device messages are sent to the devices that were used to encrypt these member events.

The schema for `m.rtc.encryption_key` to-device messages is as follows:

```json5
// PUT /_matrix/client/v3/sendToDevice/m.rtc.encryption_key/{txnId}
// Unencrypted content of OlmPayload shown, but in reality this would be an encrypted message

{
    "room_id": "{room_id}",
    "member_id": "{member_id}",
    "media_key": {
      "index": {index},
      "key": "{encoded_key}",
    },
"index": <index>,
"key": "{encoded_key}",
},
"format": 0

> }

Reviewer 13: indenting has gone badly

Reviewer: Argh, thanks. Fixed in a67a217f5cb441b05e57b8abf787b6d1098e6739.

```

- `room_id` (required, string): The ID of the room that the slot is located in.
- `member_id` (required, string): The `member.id` value of the sender's `m.rtc.member` event.
  Note that because `member.id` is unique per member, it is sufficient to disambiguate multiple
  key events for the same device.
- `media_key` (required, object): Information on the key material.
  - `key` (required, string): The key (32 bytes) encoded as specified by `format`.
  - `index` (required, number): The rolling index of the key to distinguish it from other keys. The
    value MUST be between 0 and 255 inclusive. WebRTC-based transports may use this as the `keyID`
    field of [SFrame](https://www.w3.org/TR/webrtc-encoded-transform/#sframe) headers.

> - `format` (required, number): The format in which the key was exported. Only `0` is allowed for now

Reviewer 13: We don't really use numbers anywhere else in the spec, favouring constants, eg. `m.base64`, but also are we anticipating having other formats in the future? What for? And also why at a different level to the rest of the key data?

Reviewer 11: We have a number for the megolm exchange format in the spec https://spec.matrix.org/v1.17/olm-megolm/megolm/#session-sharing-format 
Yes it is in anticipation of format change, hoping it can help transition/compatibility

Reviewer: Good point on the location of the `format` property. I agree that it makes more sense to have it next to the actual data using it and have moved it inside the `media_key` object with 44caf05abfe39dc9f87d79669a38978f10e7aac2.

The question of whether it should be a string or an integer remains though. I'd be curious why we're using integers in Megolm. I suppose it is to achieve compact packing and reduce the size of messages? @uhoreg maybe you could weigh in on this with your crypto experience?

Reviewer 10: The megolm formats predate me, but I would guess that it's an integer there because it's a binary encoding, and/or for compactness.  Those are the reasons why we use an integer in the QR code verification.
I think that here, a string (`m.base64`) would make more sense since it is more descriptive.

Reviewer: I see, thanks. Have switched to `m.base64` with 6b414dd523fec8b755ac52b4b21d76ce1bdc434a.

  and implies that the key's raw bytes were encoded using unpadded base64.

Upon receipt, clients SHOULD discard any `m.rtc.encryption_key` events that were sent in cleartext.

Receiving clients can determine the sender's `m.rtc.member` event by matching its `member.id` with
the value of `member_id` in the `m.rtc.encryption_key` message. Once the member event was determined,
clients verify that the sender and device that was used to send the member event match the sender

> and device of the to-device message. Otherwise the message MUST be discarded.

Reviewer 13: If it needs to match the sender & device, what's the purpose of the member_id?

edit: ok, I think the answer is that it disambiguates between the same device in different sessions?

Reviewer: Yes, that's it. This is also mentioned in the definition of `member_id` right above:

> [...] it is sufficient to disambiguate multiple key events for the same device.

In keeping with [MSC4153: Exclude non-cross-signed devices][MSC4153], clients SHOULD also discard

> `m.rtc.encryption_key` events when the sending device is not cross-signed.

Reviewer 13: Probably need to be a bit stricter about what exactly we mean by "cross signed" here. I think what you mean is that the device needs to be verified its user.

Reviewer 11: If we refer to the new terminology https://github.com/matrix-org/matrix-spec-proposals/blob/andybalaam/crypto-terminology/proposals/4161-crypto-terminology.md
We should say if the device is confirmed, i.e cross-signed by it's owner

Reviewer: I've added "by its owner" with 5495f4b3851aececcaf58cffd427ad2d01ddc058 to make things more clear. Otherwise, I think this is sufficiently unambiguous given that we're using the same language as https://github.com/matrix-org/matrix-spec-proposals/pull/4153 (which is referenced here and already merged).

members from decrypting past/future RTC data.

Additionally, clients SHOULD also rotate their key on a periodic schedule regardless of whether
the members have changed. This limits the impact of compromised keys.

In order to account for the delivery latency of to-device messages, clients SHOULD add a short
delay after sending a new key before starting to use it. Otherwise, receiving members may
be unable to decrypt the sender's streams temporarily. The RECOMMENDED delay duration is 5 seconds.

Furthermore, resending to-device messages to all members can be expensive when multiple
members join and/or leave in short succession. To mitigate this, clients MAY apply
some flexibility to exactly when a rotation happens relative to a membership change. This means
accepting a small window in which joining or leaving members could decrypt media that is
slightly outside their actual membership period in exchange for fewer key rotations.

As an example, let's assume a client applies the time interval `delay = 5s` between rotating a key
and starting to use it for encryption. When members leave during `delay`, the client

> could schedule another rotation for when `delay` has elapsed. This coalesces multiple member

Reviewer 13: I think 'another' is the opposite of what you mean here? Schedule a new rotation in place of the old one I think is what's intended?

Reviewer 11: No we keep the old one, but we want to coalesce several consecutive leave to a single new one at the end of the period. " Schedule a new rotation in place" if I got it right would mean that you could keep a key alive for very long if everytime there is a membership change that schedule another one later

Reviewer: Yes, exactly. This is also depicted in the chart right below. @dbkr does that help?

Reviewer: Oh, I see what you mean. Agree, that it shouldn't matter much one way or another but is needlessly ambiguous. I have tweaked the text a little in https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3853596272 to make it clear that the idea here is to collect any leavers within the 5s window and only then start a new rotation (effectively coalescing multiple leavers into one rotation). Please let us know in case this isn't better than before.

mechanisms as the remaining members at the time of leaving would have to be known ahead of time.

### Chaining member events with relations

An earlier version of this proposal used `m.reference` relations to link updated `m.rtc.member`
events to the initial joining event.

```
(Join)           (Update)              (Leave)          (Rejoin)         (Update)

m.rtc.member ──► m.rtc.member ─ ... ─► m.rtc.member     m.rtc.member ──► m.rtc.member ─ ...
      ^                │                     │                ^                │
      ├────────────────┘                     │                └────────────────┘
      │      m.reference                     │                       m.reference
      └──────────────────────────────────────┘
                                  m.reference

Time ─────────────────────────────────────────────────────────────────────────────────────►
```

This was meant to assist in reconstructing historical sessions efficiently. However, the relations
turned out to not be helpful because finding the slot as well as other members' member events
still required manual history traversal while employing timestamp overlap logic.

### Transport provisioning models

> This proposal requires the server-side Matrix deployment to also provide the MatrixRTC transport

Reviewer 13: By 'server side matrix deployment' you mean the homeserver or entity providing it? What part makes it specific to that?

Reviewer: Since transport discovery was moved to #4519, I removed this entire section in adb8fa5d2973c5681bff27e909e31b2154661f96. We'll clarify it when reintroducing it in #4519 if needed.

infrastructure. Alternatives that were considered and discarded include:

* A transport system separate from Matrix accounts – Users could obtain an account with a separate
  service provider for the RTC transport infrastructure. This is difficult to achieve across federation,
  however, since all members joined to a slot would need an account with the same external service
  provider.

> * Client-provided transports – Clients themself could define and operate transport infrastructure such as

Reviewer 13: ```suggestion
* Client-provided transports – Clients themselves could define and operate transport infrastructure such as
```

defines how these components are plugged into the system while leaving the introduction of concrete applications
and transports to other proposals.

As first concrete instances, [MSC4196] proposes a voice and video conferencing application, and
[MSC4195] proposes a transport based on the [LiveKit SFU].

[LiveKit SFU]: https://docs.livekit.io/reference/internals/livekit-sfu/
[MSC4196]: https://github.com/matrix-org/matrix-spec-proposals/pull/4196
[MSC4195]: https://github.com/matrix-org/matrix-spec-proposals/pull/4195

This proposal also doesn't cover notifications for RTC sessions. These are considered outside the core
protocol and are described in [MSC4075: MatrixRTC notifications & call ringing][MSC4075] and
[MSC4310: MatrixRTC decline m.rtc.decline][MSC4310].

[MSC4075]: https://github.com/matrix-org/matrix-spec-proposals/pull/4075
[MSC4310]: https://github.com/matrix-org/matrix-spec-proposals/pull/4310

## Proposal

### Slots

MatrixRTC slots act as virtual locations for MatrixRTC applications to run in. Slots are defined by
state events of type `m.rtc.slot`, which means that they can only be created or modified by users

> with sufficient power level. This design deliberately separates slot management from slot membership,

Reviewer 14: I'm missing a mention of how slots relate to rooms here ("room" isn't mentioned once in this paragraph).
To my understanding, and going from "Slots are represented in room state" in the introductory paragraph, a slot can only exist within an existing room, and they are in a `1 room : n slots` relationship. So the "sufficient power level" mentioned here refers to a user's power level according to the room state of the slot's room. But all of this is not made explicit.

Reviewer: Good call. Does this make it clearer?

```suggestion
MatrixRTC slots act as virtual locations for MatrixRTC applications to run in. Slots are tied to rooms
and represented by state events of type `m.rtc.slot`. This means that slots can only be created or
modified by users with sufficient power level. This design deliberately separates slot management from slot membership,
```

The fact that slots are also tied to specific applications might be clear enough through the wording a few lines further down?

> A slot is always associated with one specific application [...]

Reviewer 14: Yes, this already helps; and agreed that the `1 slot : 1 application` relationship is sufficiently clear already.

> which is introduced [below] and typically requires lower power level.

Reviewer: ```suggestion
modified by users with sufficient power level. This design deliberately separates slot management
from slot membership, which is introduced [below] and typically requires lower power level.
```

[below]: #membership

A slot is always associated with one specific application, by way of its slot ID. The slot ID is
used as the `state_key` of the `m.rtc.slot` event and is constructed as follows:

```json5
slot_id = {application_type}#{application_slot_id} (= state_key)
```

> 

Reviewer 15: There should be a sentence or brief paragraph here to state why the slot ID needs to be structured (or rather, why it can't be an opaque string). As far as I can tell, it's not actually used beyond a uniqueness guarantee - something an opaque identifier can provide. Implementations *could* use a structured form for their state keys, but they don't have to. 

Prior art here is [`m.policy.rule.*` state events](https://spec.matrix.org/v1.19/client-server-api/#mpolicyruleuser).

Reviewer 16: aa2d746c463399eb1b1d2061060358cf7db9df0d explains, hopefully. Reopen if still unclear.

> `application_type` is the application's globally unique identifier. This identifier is defined
> by the application's specification and MUST follow the [Common Namespaced Identifier Grammar].

Reviewer 15: It's somewhat easy to forget what an "application" is after all the link clicking above - a small reminder helps:
```suggestion
`application_type` is the application's globally unique identifier. This identifier is defined
by the application's specification and MUST follow the [Common Namespaced Identifier Grammar].
In the case of MSC4196, that would be `m.call`.
```

`m.rtc.slot` events have the following schema:

```json5
{
  "type": "m.rtc.slot",
  "state_key": "{application_type}#{application_slot_id}", // = slot_id
  "content": {
    "status": "{status}",
    "application": {
      "type": "{application_type}",
      ... // Further application-specific properties (if required)
    },
    "encryption": {
      "type": "{encryption_type}",
      ... // Further encryption-specific properties (if required)
    }
  },
  ...
}
```

- `status` (required, string): The slot's current status. MUST be one of `"open"`, `"closed"`.
- `application` (object): Describes the application that can run in this slot.
  - `type` (required, string): The globally unique application identifier. MUST follow the
    [Common Namespaced Identifier Grammar].
  - Optionally includes further properties for settings that are specific to the application
    `type`. The concrete properties are defined by the application's specification. A calling
    application, for instance, could include properties for constraining the call to be voice-only.
- `encryption` (object): Describes the encryption mechanism to use in this slot. Further,

>   details on the available mechanisms can be found in the [encryption section] below.

Reviewer 14: [This conversation](https://github.com/matrix-org/matrix-spec-proposals/pull/4143#discussion_r3629100442) made me realise that it's never made explicit (I think) what causes encryption to be considered as enabled or disabled, for example for the paragraph in the [End-to-End encryption section](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#end-to-end-encryption) containing things like 
> once encryption is enabled on a slot, it can never be disabled

I suppose encryption is considered to be _enabled_ if the optional `encryption` object is present, and considered _disabled_ if it's absent? I think this should be mentioned here.

Reviewer: Err, yes indeed. 🤦‍♂️ 

```suggestion
- `encryption` (object): If present, describes the encryption mechanism to use in this slot. Further,
  details on the available mechanisms can be found in the [encryption section] below. If absent,
  encryption is disabled.
```

  - `type` (required, string): The globally unique identifier of the encryption mechanism.
    MUST follow the [Common Namespaced Identifier Grammar].
  - Optionally includes further properties for settings that are specific to the encryption
  `type`.

[encryption section]: #end-to-end-encryption

#### Slot lifecycle

A slot is opened by sending an `m.rtc.slot` state event with `status = "open"`, a valid `application`
object that aligns with the slot's `state_key` and, if needed, a valid `encryption` object. To close a
slot, the corresponding `m.rtc.slot` state event is updated with `status = "closed"`. The `application`
and `encryption` objects are not required on closed slots but may be kept around for convenience to

> simplify re-opening the slot. Slot events that don't match the schema above, SHOULD be considered closed.

Reviewer 14: > Slot events that don't match the schema above, SHOULD be considered closed.

The wording is a bit weird here, is this supposed to mean: If there's an `m.rtc.slot` state event with a `state_key` of `slot_id` that doesn't match the schema above, then the slot with that `slot_id` SHOULD be considered closed?

Reviewer: Yes, but not only that. "schema" was actually supposed to mean the JSON schema starting on line 87 above (which includes the `state_key` condition.

Does this help?

```suggestion
simplify re-opening the slot. Slot events that don't match the JSON schema above, SHOULD be considered closed.
```

Reviewer 14: Hm, I don't think this does it for me yet.

The part I find weird is that the sentence in effect says that "Slot events [are] considered closed", but I don't think an _event_ being considered closed makes much sense – it's really _slots_ that are open/closed. So if there's no (well-formed) `state_key`, what is it that should even be considered closed?

And for making this explicitly about the JSON schema: For example, what if
> a valid `application` object that aligns with the slot's `state_key`

 which is only mentioned as an additional requirement in prose isn't met?

Reviewer: Ok yes, I see. I have made the requirements a little more explicit in the schema itself with https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3629860647 and https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3629888299.

How about we tighten the wording in this paragraph like so?

> A slot is opened by sending an `m.rtc.slot` state event with `status = "open"`, a valid `application`
> object and, if needed, a valid `encryption` object as per the JSON schema above. Any slot that
> doesn't fulfill these requirements is closed. To explicitly close an open slot, the associated `m.rtc.slot` > state event is updated with `status = "closed"`. The `application` and `encryption` objects are not
> required on closed slots but may be kept around for convenience to simplify re-opening the slot.

Reviewer 14: Both the explicit requirements and the new wording look good to me, thanks!
(note that there's a rogue `>` in your paragraph in "`m.rtc.slot` > state event")

Reviewer: > (note that there's a rogue > in your paragraph in "m.rtc.slot > state event")

Argh, thanks. 🤦‍♂️ 

Fixed in https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3630188997.

state event is updated with `status = "closed"`. The `application` and `encryption` objects are not
required on closed slots but may be kept around for convenience to simplify re-opening the slot.
The semantics of open and closed slots for actual slot membership are described in the membership event
section [below].

Slots may follow different lifecycles depending on the use case. For instance, a long-lived slot
that is kept open continually could power a Discord-style experience where members can hop on
and hop off as desired. Scheduled conference meetings, in turn, could benefit from a time-bounded
slot that is only opened when the meeting starts and closed again afterwards.

For the time being, slots will have to be created manually. A future proposal may change the

> defaults for newly created rooms to provide slots for standard RTC applications.

Reviewer 15: In the interest of making the MSC slightly fewer lines:
```suggestion
Slots do not close automatically. A slot that is deliberately long-lived could be used to
create a Discord-style experience where members hop on and off as desired. A conferencing
or meeting-centric client might close its slots at the end of the meeting. A future
proposal might change how slots are automatically managed.
```

    },
    "transports": {
      "published": [
        {
          "type": "{transport_type}",
          ... // Further transport-specific properties (if required)
        },
        ...
      ],
      "can_subscribe": [
        "{transport_type}",
        ...
      ]
    },
    "sticky_key": "{member_id}" // = member.id
  },
  ...
}
```

- `slot_id` (required, string): The `state_key` of the slot that is being joined.
- `application` (required, object): Describes the application that is running in the slot.
  - `type` (required, string): The application's globally unique identifier; same as in `m.rtc.slot`.
  - Optionally includes further properties for settings that are specific to the application
    `type`. As in `m.rtc.slot`, the concrete properties are defined by the application's specification.
    For example, a [Third Room](https://thirdroom.io) application could include approximate map positions,
    allowing clients to avoid connecting to members outside their area of interest.
- `member` (required, object): Information to identify the member.
  - `id` (required, string): Identifier to distinguish multiple members, even for the same user

>     and device. MUST be unique for each join. This means that clients need to use a different

Reviewer 14: Who is responsible for ensuring this "MUST" requirement, and how are they supposed to do that? What happens / should happen if someone sends an `m.rtc.member` event with an already existing `member.id`?

Reviewer: The sticky map algorithm includes this ID and the event sender in the composite map key. So uniqueness is only required among members / devices of the same user.

Does this help clarify that?

```suggestion
    and device. MUST be unique for each join of the same user. This means that clients need to use a different
```

If the same user, happens to use identical IDs on two different joins, that would make the membership events collide in the sticky map. However, I think they could only attack themselves.

Reviewer 14: Okay, yes, this does clarify it. The version without "of the same user" read to me like member IDs were supposed to be unique even among several users.

>       [Common Namespaced Identifier Grammar] but without the namespacing requirements.

Reviewer 15: Like above: a small example helps clarify terminology.
```suggestion
      [Common Namespaced Identifier Grammar] but without the namespacing requirements. For a
      MSC4195 transport, this would be `livekit`.
```

Reviewer 11: BTW @turt2live , the [MSC4519](https://github.com/matrix-org/matrix-spec-proposals/blob/travis/msc/voip-transports-registry/proposals/4519-rtc-transports-registry.md) is saying the type is an Opaque Identifier. 
Should we update here? A Namespaced identifier without the namespacing requirement is just an Opaque identifier anyhow?
Or maybe we should namespace

Reviewer 14: As I happened to wonder and read about this recently: [the spec says](https://spec.matrix.org/v1.16/appendices/#common-namespaced-identifier-grammar)
> “this identifier uses the Common Namespaced Identifier Grammar, though without the namespacing requirements” - this means that m. is still reserved, but that implementations do not have to use the reverse DNS scheme to namespace their custom identifier.

and also `~` and `[A-Z]` are allowed for Opaque Identifiers but not for Namespaced Identifiers – so there are some differences.

Reviewer 16: 00c0b07c1f0ce17282a5187da76b384a7faf0a6c

Reviewer 16: > Should we update here? A Namespaced identifier without the namespacing requirement is just an Opaque identifier anyhow? Or maybe we should namespace

https://github.com/matrix-org/matrix-spec-proposals/pull/4520 moves the problem of deciding the identifier grammar out to MSC4519, so I've continued discussion there: https://github.com/matrix-org/matrix-spec-proposals/pull/4519#discussion_r3682414975

    identifier when leaving and then rejoining a slot.
  - `claimed_device_id` (required, string) — Matrix device identifier. This is used to exchange
    encryption keys as explained later in the [encryption section]. The device ID is untrusted ("claimed"

>     by the sender) and must be cross checked against the message encryption envelope for confirmation.

Reviewer 14: What exactly must the `claimed_device_id` be checked against, and what exactly does the "message encryption envelope" refer to – is it supposed to be checked against the `device_id` in the `content` of an Megolm-encrypted `m.room.encrypted` event, which is defined in [this section](https://spec.matrix.org/v1.19/client-server-api/#mmegolmv1aes-sha2)?

But what if encryption isn't enabled for the room?

Reviewer: > What exactly must the `claimed_device_id` be checked against, and what exactly does the "message encryption envelope" refer to – is it supposed to be checked against the `device_id` in the `content` of an Megolm-encrypted `m.room.encrypted` event, which is defined in [this section](https://spec.matrix.org/v1.19/client-server-api/#mmegolmv1aes-sha2)?

I _think_ so. @BillCarsonFr can you confirm that?

> But what if encryption isn't enabled for the room?

The verification is only required when `claimed_device_id` is used for exchanging encryption keys. That should only happen in encrypted rooms.

Reviewer 14: > The verification is only required when claimed_device_id is used for exchanging encryption keys. That should only happen in encrypted rooms.

But one could enable encryption for MatrixRTC even in rooms that are not themselves encrypted, right? I'm not sure how useful this is, but if it is indeed possible, I think it should be specified what's supposed to happen in that case.

Also, the check only being required in some cases is not what's currently written down: `claimed_device_id` is marked as required, and it "must be cross checked" as per the current wording – but that's just not possible for non-encrypted rooms if my interpretation from above is correct.

Reviewer: Hold on, I think the verification isn't performed against the encryption envelope on the `m.rtc.member` event but against the device ID from the Olm payload. That always exists even if using encryption in an unencrypted room.

Reviewer 11: > What exactly must the claimed_device_id be checked against
> ..
>  is it supposed to be checked against the device_id in the content of an Megolm-encrypted

First the `device_id` in the megolm message is deprecated https://github.com/matrix-org/matrix-spec-proposals/pull/3700, so it should not be used anymore.

But that is sort of what you want to do. When a megolm encrypted message is received, it is decrypted by a megolm key that was received via an olm-channel. When received via olm-channel the megolm key is then linked to the sending device. That is what you want to check, that the device that is cryptographicaly linked to the megom message is the same as the one that is claimed in the membership content.

So as you said, in a clear room there is no way to check if it is claimed or not (but in a clear room, your message can also be spoofed by the homeserver admin).

> But one could enable encryption for MatrixRTC even in rooms that are not themselves encrypted, right?

Yes, the call can be encrypted but it doesn't mean that the keys are distributed also encrypted. I suppose it is for clients that cannot do encryption?

@Johennes Maybe we need to clarify and set the expectation in a clear room?

There is also an alternative where we could just remove the `claimed_device_id` and just use the broadcast wildcard `*` to send the key to the user. But it will cost bandwith

Reviewer 14: > But that is sort of what you want to do. When a megolm encrypted message is received, it is decrypted by a megolm key that was received via an olm-channel. When received via olm-channel the megolm key is then linked to the sending device. That is what you want to check, that the device that is cryptographicaly linked to the megom message is the same as the one that is claimed in the membership content.

Sorry, I still don't fully understand.
- Are you saying the `claimed_device_id` *is* checked against the `device_id` in the megolm envelope, even though it is deprecated?
- Or are you saying `claimed_device_id` is checked against some internally stored `device_id` that is cryptographically linked to the Megolm session through which the `m.rtc.member` event was received, and which originally stems from the Olm-encrypted to-device message that was used to send the Megolm key? (If that's the case, this is not at all obvious from the MSC.)

Also, I'm still trying to wrap my had around what this check is even achieving, I think I need to take a bit more time to really understand this.

> So as you said, in a clear room there is no way to check if it is claimed or not ... Maybe we need to clarify and set the expectation in a clear room?

Yes, I do think so, because, as I mention above, the current wording implies that a check has to happen under all circumstances.

> Yes, the call can be encrypted but it doesn't mean that the keys are distributed also encrypted.

This made me take a look at the _Distributing keys_ section again, where I noticed
"Upon receipt, clients SHOULD discard any `m.rtc.encryption_key` events that were sent in cleartext."
so... how does that fit together?

> There is also an alternative where we could just remove the claimed_device_id and just use the broadcast wildcard * to send the key to the user. But it will cost bandwith

I don't follow this, do you suggest to remove the `claimed_device_id` overall, or just in the case where the room is not encrypted? And where would "the broadcast wildcard *" be used in its stead?

Generally, the _unencrypted room – encrypted RTC_ and the opposite _encrypted room – unencrypted RTC_ combinations seem to be not very well defined to me, and might need some explicit mentioning. (Also see this other thread: https://github.com/matrix-org/matrix-spec-proposals/pull/4143#discussion_r3623554246)

Reviewer: This should be resolved once https://github.com/matrix-org/matrix-spec-proposals/pull/4516 lands.

Reviewer 11: Resolved as #4516 has been merged

  - `can_subscribe` (array): An array of transport types that the member is able to subscribe to.
    Other members can use this as cue for deciding which transports to use to accommodate this member.
- `sticky_key` (required, string): The sticky key for the ephemeral map algorithm as defined
  in the addendum of [MSC4354]. MUST have the same value as `member.id`.

Apart from having to match the above schema, an `m.rtc.member` event SHOULD only be considered to be
joined if all of the following conditions apply:

- An open slot exists in the room state as an `m.rtc.slot` state event with `state_key` equalling
  the `m.rtc.member` event's `slot_id`.
- The sender is currently a member of the room (i.e. has membership `join`).
- The event is currently sticky, meaning that its stickiness duration as per [MSC4354] has not expired.
  This is to ensure that the membership view is as consistent as possible across all members.
- If the room is encrypted, the `m.rtc.member` event was sent encrypted rather than in clear.

If these conditions are not fulfilled, clients SHOULD treat the member as left and refrain

> from sending them encryption keys or connecting to their transports.

Reviewer 14: > an `m.rtc.member` event SHOULD only be considered to be joined ...

> clients SHOULD treat the member as left ...

https://github.com/matrix-org/matrix-spec-proposals/pull/4510/changes apparently changed the two MUST occurrences to SHOULD here, why is that? (I couldn't find a reason, if this was already discussed somewhere, let me know.)
Especially the second and fourth condition seem important if we want to avoid unauthorised users to receive `m.rtc.encryption_key` messages, see
> The recipient devices are determined from the `m.rtc.member` events that are considered to be joined to the slot.

[below](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#distributing-keys), which makes me feel like this should be a MUST requirement.

Reviewer: The thinking was that a paranoid client may want to present users which have recently left the session as still being part of the session for the duration of the key rotation grace period, so the user is aware of the potential for eavesdropping. Does that make sense?

Reviewer 14: Hm, yes, this makes sense, but I still think for the purpose of distributing encryption keys, the MUST version is more appropriate, especially since the encryption section explicitly references these conditions with
> The recipient devices are determined from the `m.rtc.member` events that are considered to be joined to the slot. The conditions for considering a member joined were given [above](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#joining-a-slot).

I guess the problem is that "a paranoid client" would really want two differing membership definitions for two different purposes:
1. Who might potentially receive call data and decrypt it, i.e., should be presented to the user as still joined? This is why the SHOULD could be considered appropriate.
2. Who should receive (new) encryption keys? This is always a subset of case (1) and in my opinion needs the MUST.

I'm not sure how this is best described in the MSC without overcomplicating everything though.

Reviewer: I see what you mean. I actually think we can switch to MUST here without denying the paranoid behavior. A client can "treat the member left" (e.g. by blurring their tile) and still visualize that they may still have decryption keys due to the key rotation optimization.

Reviewer: Fixed in https://github.com/matrix-org/matrix-spec-proposals/pull/4143/commits/3e3df5a0e9a034ff24091fa1ad523e29de6c3342 and https://github.com/matrix-org/matrix-spec-proposals/pull/4143/commits/2c1ed9eb68ed0ee4d86804ee048df7e60a26f569.

      "membership": "leave",
    },
    "leave_reason": {
      "code": "{code}",
      "reason": "{reason}",
    },
    "sticky_key": "{member_id}" // = member.id from previously joined m.rtc.member event
  },
  ...
}
```

Again, once a member has left, clients SHOULD refrain from connecting to their transports.

[lifecycle]: #membership-lifecycle

#### Membership lifecycle

A typical lifecycle of a MatrixRTC membership involves a series of `m.rtc.member` events, as follows:

1. members first join a slot by sending a joining `m.rtc.member` event.
1. Afterwards, members may update their membership, e.g. to change transports or modify
   application-specific settings, by sending a new `m.rtc.member` event with the same `sticky_key`.
   Since the actual join state is constrained by the stickiness of the member event, clients
   also need to send new `m.rtc.member` events if they want to stay joined for longer than the stickiness
   duration. It is RECOMMENDED that clients send these events sufficiently ahead of the stickiness
   expiration to minimize potential join state flickering.
1. Finally, to leave the slot, members send a leaving `m.rtc.member` event.

> As explained above, the resolved membership state is also constrained by the associated `m.rtc.slot`

Reviewer 11: I think this has a side effect for sliding sync, the `m.rtc.slot` must be added in the list of `required_state` config of MSC4186. Should this be documented here? 
If it is not in required_state, the client should query the state before validating memberships

Reviewer 16: The `required_state` is always determined by the client, right? I don't see mention of specified defaults, so IMO it's fine to leave implied that a MatrixRTC client using SSS would want to do something to get the `m.rtc.slot` state.

Reviewer 11: Added as an implementation task in the project board, resolving.

in other members considering the member as still joined, possibly for longer periods,
even though no media can be exchanged. To mitigate the impact of this, clients SHOULD use
delayed events as per [MSC4140: Cancellable delayed events][MSC4140] to implement a "dead man's switch".
This means scheduling the `m.rtc.member` leave event as a delayed event with a reasonably
short delay (e.g. 15-30 seconds). While being connected, the client can periodically restart
the delayed event to push it into the future. If the client then happens to lose connectivity
and the delay times out, the homeserver will automatically send the leaving `m.rtc.member` event.

[MSC4140]: https://github.com/matrix-org/matrix-spec-proposals/pull/4140

### Sessions

MatrixRTC sessions only exist indirectly through the temporal overlap of `m.rtc.member` events
that are considered to be joined to the same `m.rtc.slot` event. In other words, a session
represents the span of time during which a potentially changing set of one or more members
is continuously joined to the same slot.

The examples below illustrate how different membership lifecycles and slot configurations
give rise to sessions.

```

> m.rtc.member[0]            |████████████████████████            ████████████████████████|████

Reviewer 14: I'm confused, why does the membership extend beyond the leave and the session end?

Should this actually be

```suggestion
m.rtc.member[0]            |████████████████████████            ████████████████████████|
```

Reviewer: This is actually not the computed membership but just the state of the `m.rtc.member` event. In this example, the computed membership changes to left when the slot is closed but the user only updates their membership events afterwards (which doesn't affect the computed membership). Maybe the `leave ^` should point to the end of the `m.rtc.member` block to make that clearer?

Reviewer 14: At least for me, that would make it clearer, as I interpreted the `leave` to mean that an according `m.rtc.member` was sent.

Reviewer: Fixed in https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3630023360

  multiple transports interchangeable (e.g. when advertising multiple transports of the same type),
  it SHOULD arrange them in descending order of preference (e.g. listing backup infrastructure last).
  - `type`: (required, string): The globally unique transport identifier. MUST follow the
    [Common Namespaced Identifier Grammar] but without the namespacing requirements.
  - Optionally includes further properties specific to the transport `type`. The concrete properties
    are defined by the transport's specification.

### End-to-end encryption

`m.rtc.member` events MUST be encrypted when sent in an encrypted room.

Additionally, applications need a way to encrypt the RTC data itself. The process is generally
specific to the transport being used, but often requires session members to agree on key material, at a minimum.
To support this, MatrixRTC provides a generic system for establishing shared key material between
members. Transports can then define how to actually use this key material, which may involve
deriving further secrets from it.

The concrete mechanism for agreeing on the shared key material within a slot is prescribed through

> the `encryption` object in `m.rtc.slot` events. Clients SHOULD enforce the use of encryption when

Reviewer 14: What does it mean for a client to "enforce" encryption? Should it just ignore `m.rtc.slot` events that do not enable encryption, and not make use of the according slot? Or should it try to use encryption for the slot regardless of what the `m.rtc.slot` says somehow?

Reviewer: The statement here is only concerned with opening a slot actually. Enforcing the use of encryption when opening a slot would mean that a client doesn't allow the user to open the slot without encryption if the room itself is also encrypted. Does that help?

Reviewer 14: So if a client receives an `m.rtc.slot` event in an encrypted room that doesn't enable encryption for the slot, it accepts it (as long as encryption wasn't ever enabled for the slot) and continues without using any encryption mechanism for the RTC data?

Reviewer: Yes, that is basically another slightly odd but technically possible combination. An unencrypted RTC session in an encrypted room. We could require clients to refuse connecting to such slots. But then I think the spec doesn't prohibit unencrypted events in encrypted rooms either? 🤔

Reviewer: This should be resolved once https://github.com/matrix-org/matrix-spec-proposals/pull/4516 lands.

Reviewer 11: resolved as #4516 did land

distributed among session members. Other devices, even if in the room, never get the key material.

[to-device messages]: https://spec.matrix.org/v1.18/client-server-api/#send-to-device-messaging

#### Distributing keys

When joining a slot, clients generate a 32-byte cryptographically secure key. They then share
the key with other clients joined to the slot by sending encrypted to-device messages of the
type `m.rtc.encryption_key`.

The recipient devices are determined from the `m.rtc.member` events that are considered to be
joined to the slot. The conditions for considering a member joined were given
[above](#joining-a-slot). Once joined members are determined, the target device ID
is taken from the `member.claimed_device_id` property of the respective `m.rtc.member` event.

```json5
// PUT /_matrix/client/v3/sendToDevice/m.rtc.encryption_key/{txnId} 
// (Unencrypted contents shown, but in reality this would be an encrypted message)

{
    "room_id": "{room_id}",
    "member_id": "{member_id}",
    "media_key": {
      "index": {index},
      "key": "{encoded_key}",
    },
    "format": "0"
}

> ```

Reviewer 14: With this being an Olm-encrypted to-device message, the JSON here is probably the `content` of an [`OlmPayload`](https://spec.matrix.org/v1.19/client-server-api/#molmv1curve25519-aes-sha2) (now called `OlmPlaintext`, see [this PR](https://github.com/matrix-org/matrix-spec/pull/2418)), is that right? I think this should be mentioned, given it is important for understanding the bullet points I comment on below.

Reviewer: Yes, good point. Fixed in https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3630063576 and https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3630072694.

Reviewer 14: Unfortunately I don't think that's quite right, as it's only the `content` of the `OlmPlaintext`, not the whole `OlmPlaintext` that's shown here – or so I believe.

Reviewer: This should be resolved once https://github.com/matrix-org/matrix-spec-proposals/pull/4516 lands.

  key events for the same device.
- `media_key` (required, object): Information on the key material.
  - `key` (required, string): The key (32 bytes) encoded as specified by `format`.
  - `index` (required, number): The rolling index of the key to distinguish it from other keys. The
    value MUST be between 0 and 255 inclusive. WebRTC-based transports may use this as the `keyID`
    field of [SFrame](https://www.w3.org/TR/webrtc-encoded-transform/#sframe) headers.
- `format` (required, number): The format in which the key was exported. Only `0` is allowed for now
  and implies that the key's raw bytes were encoded using unpadded base64.

Upon receipt, clients SHOULD discard any `m.rtc.encryption_key` events that were sent in cleartext.

Receiving clients can determine the corresponding `m.rtc.member` event by matching its `member.id`
with the value of `member_id` in the `m.rtc.encryption_key` message. Once the member event was determined,
clients perform the following checks:

- The `sender` property from the decryption result matches the `sender` of the `m.rtc.member`
  event.
- The `device_id` property from the decryption result matches the `member.claimed_device_id`

>   value of the `m.rtc.member` event.

Reviewer 14: What is the "decryption result" referring to? My first assumption would be the plaintext JSON that is given above (which I commented on), but that doesn't have any `sender` or `device_id` property. So it's probably referring to the decrypted [`OlmPayload`](https://spec.matrix.org/v1.19/client-server-api/#molmv1curve25519-aes-sha2) (/`OlmPlaintext`), and `device_id` is the `sender_device_keys.device_id`?

Reviewer: This should be resolved once https://github.com/matrix-org/matrix-spec-proposals/pull/4516 lands.

document. This exposes transports to unauthenticated users, however, which can be a security concern.
Additionally, in enterprise deployments, `.well-known` files are often not served by the homeserver itself
and it can be bureaucratically complicated to update entries under the top-level domain.

`GET /_matrix/client/v1/rtc/transports` avoids these issues and offers more flexibility for future extensions
such as user-specific transports.

### Key distribution via room events

Earlier iterations of this MSC used encrypted room events to distribute per-member encryption
keys. This turned out to be problematic due to homeservers rate-limiting message sending, timelines
being polluted with invisible events and, most importantly, the keys being shared with all room
members rather than just the session members.

#### Shared key encryption

For large calls an encryption scheme based on a shared key instead of per-sender keys could be more
efficient. This would obviously weaken security properties though. A future proposal may consider the
tradeoff and introduce a shared-key system via a new encryption `type`.

> ## Security considerations

Reviewer: @gewitternacht remarked out of band:

> The current slot membership is determined by clients only (not the server), for example by checking whether the slot given in an m.rtc.member event is actually still open. So if two clients agreed to engage in some shenanigans, they could both send some m.rtc.member events and act as if the according slot is open even though the latest m.rtc.slot state event defines that slot as closed.
> Or more generally, if some clients agreed that they don't look at m.rtc.slot events anymore but instead use some alternative m.rtc.slot.alt event type that they invented, they could use MatrixRTC just fine, but without the power-level restriction that guards m.rtc.slot events.

We discussed this in chat and acknowledged the possibility but couldn't see any obvious issues with such "shadow sessions". Nevertheless, this section should document that we've thought about it.

Reviewer 14: Adding why I think it's important to document this: It means that future applications _must not_ depend on no one using a slot in any way if it is closed – which might be the expectation and is suggested by the `m.rtc.slot` events, but cannot be guaranteed as per the above.
Still, e.g. for MSC4196, we don't think this is currently a problem in practice, and such non-conforming clients shouldn't affect the experience of conforming clients.

Reviewer: How about this?

```suggestion
## Security considerations

### Shadow sessions

Malicious clients could ignore slot events and consider members to be joined regardless of slot
restrictions. This would allow them to engage in "shadow sessions" in the room that would be ignored
by other, conforming clients in the room. While this undermines the authoritativeness of slots, it
does not impact conforming clients in any way. RTC application and transport designers should be
aware that slots don't provide access control, however, and adopt appropriate measures.
```

Reviewer 14: This captures the points I wanted to make very well, thanks!

The grammar for forming slot IDs MUST NOT be used to parse the components out of a slot ID.
It exists only to namespace the `state_key`, and could be modified in a future proposal.

[Common Namespaced Identifier Grammar]: https://spec.matrix.org/v1.16/appendices/#common-namespaced-identifier-grammar

[^nohash]: Note that due the use of the [Common Namespaced Identifier Grammar](https://spec.matrix.org/v1.16/appendices/#common-namespaced-identifier-grammar),
neither `application_type` nor `application_slot_id` can contain the `#` character.

`m.rtc.slot` events have the following schema:

```json5
{
  "type": "m.rtc.slot",
  "state_key": "{application_type}#{application_slot_id}", // = slot_id
  "content": {
    "status": "{status}",
    "application": {
      "type": "{application_type}",
      ... // Further application-specific properties (if required)
    },
    "encryption": {
      "type": "{encryption_type}",
      ... // Further encryption-specific properties (if required)
    }
  },
  ...
}
```

- `status` (required, string): The slot's current status. MUST be one of `"open"`, `"closed"`.

> - `application` (object): Describes the application that can run in this slot.

Reviewer: ```suggestion
- `application` (object): Describes the application that can run in this slot. REQUIRED if
  `status = open`.
```

  - `type` (required, string): The globally unique application identifier. MUST follow the

>     [Common Namespaced Identifier Grammar].

Reviewer: Making the required relation between `application.type` and `state_key` more explicit as per https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3621212848.

```suggestion
  - `type` (required, string): The globally unique application identifier. MUST follow the
    [Common Namespaced Identifier Grammar]. MUST align with the event's `state_key`.
```

- `encryption` (object): If present, describes the encryption mechanism to use in this slot. Further,
  details on the available mechanisms can be found in the [encryption section] below. If absent,
  encryption is disabled.
  - `type` (required, string): The globally unique identifier of the encryption mechanism.
    MUST follow the [Common Namespaced Identifier Grammar].
  - Optionally includes further properties for settings that are specific to the encryption
  `type`.

[encryption section]: #end-to-end-encryption

#### Slot lifecycle

A slot is opened by sending an `m.rtc.slot` state event with `status = "open"`, a valid `application`
object that aligns with the slot's `state_key` and, if needed, a valid `encryption` object. To close a
slot, the corresponding `m.rtc.slot` state event is updated with `status = "closed"`. The `application`
and `encryption` objects are not required on closed slots but may be kept around for convenience to
simplify re-opening the slot. Slot events that don't match the schema above, SHOULD be considered closed.
The semantics of open and closed slots for actual slot membership are described in the membership event

> section [below].

Reviewer: From https://github.com/matrix-org/matrix-spec-proposals/pull/4143#discussion_r3629906543.

```suggestion
A slot is opened by sending an `m.rtc.slot` state event with `status = "open"`, a valid application
object and, if needed, a valid encryption object as per the JSON schema above. Any slot that
doesn't fulfill these requirements is closed. To explicitly close an open slot, the associated `m.rtc.slot`
state event is updated with `status = "closed"`. The `application` and `encryption` objects are not
required on closed slots but may be kept around for convenience to simplify re-opening the slot.
The semantics of open and closed slots for actual slot membership are described in the membership event
section [below].
```

  joined to the session. The exact procedure for publishing and subscribing to real-time data is
  defined in each transport's specification.
  - `published` (array): An array of objects describing the transports on which the member is
    publishing media.
    - `type`: (required, string): The globally unique transport identifier. MUST follow the
      [Common Namespaced Identifier Grammar] but without the namespacing requirements.
    - Optionally includes further properties specific to the transport `type`. The concrete properties
      are defined by the transport's specification. This could, for instance, include WebSocket URLs.
  - `can_subscribe` (array): An array of transport types that the member is able to subscribe to.
    Other members can use this as cue for deciding which transports to use to accommodate this member.
- `sticky_key` (required, string): The sticky key for the ephemeral map algorithm as defined
  in the addendum of [MSC4354]. MUST have the same value as `member.id`.

Apart from having to match the above schema, an `m.rtc.member` event SHOULD only be considered to be

> joined if all of the following conditions apply:

Reviewer: From https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3622626629.

```suggestion
Apart from having to match the above schema, an `m.rtc.member` event MUST only be considered to be
joined if all of the following conditions apply:
```

- An open slot exists in the room state as an `m.rtc.slot` state event with `state_key` equalling
  the `m.rtc.member` event's `slot_id`.
- The sender is currently a member of the room (i.e. has membership `join`).
- The event is currently sticky, meaning that its stickiness duration as per [MSC4354] has not expired.
  This is to ensure that the membership view is as consistent as possible across all members.
- If the room is encrypted, the `m.rtc.member` event was sent encrypted rather than in clear.

If these conditions are not fulfilled, clients SHOULD treat the member as left and refrain

> from sending them encryption keys or connecting to their transports.

Reviewer: From https://github.com/matrix-org/matrix-spec-proposals/pull/4143#discussion_r3622626629.

```suggestion
If these conditions are not fulfilled, clients MUST treat the member as left and refrain
from sending them encryption keys or connecting to their transports.
```

even though no media can be exchanged. To mitigate the impact of this, clients SHOULD use
delayed events as per [MSC4140: Cancellable delayed events][MSC4140] to implement a "dead man's switch".
This means scheduling the `m.rtc.member` leave event as a delayed event with a reasonably
short delay (e.g. 15-30 seconds). While being connected, the client can periodically restart
the delayed event to push it into the future. If the client then happens to lose connectivity
and the delay times out, the homeserver will automatically send the leaving `m.rtc.member` event.

[MSC4140]: https://github.com/matrix-org/matrix-spec-proposals/pull/4140

### Sessions

MatrixRTC sessions only exist indirectly through the temporal overlap of `m.rtc.member` events
that are considered to be joined to the same `m.rtc.slot` event. In other words, a session
represents the span of time during which a potentially changing set of one or more members
is continuously joined to the same slot.

The examples below illustrate how different membership lifecycles and slot configurations
give rise to sessions.

```
m.rtc.member[0]            |████████████████████████            ████████████████████████|████

>                            |^ join           leave ^            ^ join           leave ^|

Reviewer: From https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3623490269.

```suggestion
                           |^ join           leave ^            ^ join                  |   ^ leave
```

```json5
{
  "type": "m.rtc.slot",
  "content": {
    "encryption": {
      "type": "m.per_member"
    },
    ...
  },
  ...
}
```

Under `m.per_member` every member maintains a unique sender key. This key is shared securely
with other members via Olm-encrypted [to-device messages]. This ensures that keys are only
distributed among session members. Other devices, even if in the room, never get the key material.

[to-device messages]: https://spec.matrix.org/v1.18/client-server-api/#send-to-device-messaging

#### Distributing keys

When joining a slot, clients generate a 32-byte cryptographically secure key. They then share
the key with other clients joined to the slot by sending encrypted to-device messages of the
type `m.rtc.encryption_key`.

The recipient devices are determined from the `m.rtc.member` events that are considered to be
joined to the slot. The conditions for considering a member joined were given
[above](#joining-a-slot). Once joined members are determined, the target device ID
is taken from the `member.claimed_device_id` property of the respective `m.rtc.member` event.

```json5
// PUT /_matrix/client/v3/sendToDevice/m.rtc.encryption_key/{txnId} 

> // (Unencrypted contents shown, but in reality this would be an encrypted message)

Reviewer: From https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3623692319.

```suggestion
// (Unencrypted OlmPlaintext shown, but in reality this would be an encrypted message)
```

> // (Unencrypted OlmPlaintext shown, but in reality this would be an encrypted message)

Reviewer: From https://github.com/matrix-org/matrix-spec-proposals/pull/4143/changes#r3623692319

```suggestion
// Unencrypted OlmPlaintext (formerly OlmPayload) shown, but in reality this would be an encrypted message
```
